--- /tmp/tmp.tYSqORsSZN/ubuntu	2026-04-30 18:36:52.046060972 +0200
+++ /home/darren/development/enfs/vendor/openeuler/fs/nfs/super.c	2026-04-30 17:42:52.000000000 +0200
@@ -66,14 +66,13 @@
 #include "callback.h"
 #include "delegation.h"
 #include "iostat.h"
-#include "internal.h"
+#include "enfs_adapter.h"
 #include "fscache.h"
 #include "nfs4session.h"
 #include "pnfs.h"
 #include "nfs.h"
 #include "netns.h"
 #include "sysfs.h"
-#include "nfs4idmap.h"
 
 #define NFSDBG_FACILITY		NFSDBG_VFS
 
@@ -131,7 +130,7 @@
 }
 #endif /* CONFIG_NFS_V4_2 */
 
-static struct shrinker *acl_shrinker;
+static struct shrinker_v2 *acl_shrinker;
 
 /*
  * Register the NFS filesystems
@@ -212,14 +211,15 @@
 }
 EXPORT_SYMBOL_GPL(nfs_sb_deactive);
 
-int nfs_client_for_each_server(struct nfs_client *clp,
-		int (*fn)(struct nfs_server *server, void *data), void *data)
+static int __nfs_list_for_each_server(struct list_head *head,
+		int (*fn)(struct nfs_server *, void *),
+		void *data)
 {
 	struct nfs_server *server, *last = NULL;
 	int ret = 0;
 
 	rcu_read_lock();
-	list_for_each_entry_rcu(server, &clp->cl_superblocks, client_link) {
+	list_for_each_entry_rcu(server, head, client_link) {
 		if (!(server->super && nfs_sb_active(server->super)))
 			continue;
 		rcu_read_unlock();
@@ -238,6 +238,13 @@
 		nfs_sb_deactive(last->super);
 	return ret;
 }
+
+int nfs_client_for_each_server(struct nfs_client *clp,
+		int (*fn)(struct nfs_server *, void *),
+		void *data)
+{
+	return __nfs_list_for_each_server(&clp->cl_superblocks, fn, data);
+}
 EXPORT_SYMBOL_GPL(nfs_client_for_each_server);
 
 /*
@@ -446,12 +453,8 @@
 		{ NFS_MOUNT_NONLM, ",nolock", "" },
 		{ NFS_MOUNT_NOACL, ",noacl", "" },
 		{ NFS_MOUNT_NORDIRPLUS, ",nordirplus", "" },
-		{ NFS_MOUNT_FORCE_RDIRPLUS, ",rdirplus=force", "" },
 		{ NFS_MOUNT_UNSHARED, ",nosharecache", "" },
 		{ NFS_MOUNT_NORESVPORT, ",noresvport", "" },
-		{ NFS_MOUNT_NETUNREACH_FATAL,
-		  ",fatal_neterrors=ENETDOWN:ENETUNREACH",
-		  ",fatal_neterrors=none" },
 		{ 0, NULL, NULL }
 	};
 	const struct proc_nfs_info *nfs_infop;
@@ -515,16 +518,8 @@
 	else
 		nfs_show_nfsv4_options(m, nfss, showdefaults);
 
-	if (nfss->options & NFS_OPTION_FSCACHE) {
-#ifdef CONFIG_NFS_FSCACHE
-		if (nfss->fscache_uniq)
-			seq_printf(m, ",fsc=%s", nfss->fscache_uniq);
-		else
-			seq_puts(m, ",fsc");
-#else
+	if (nfss->options & NFS_OPTION_FSCACHE)
 		seq_puts(m, ",fsc");
-#endif
-	}
 
 	if (nfss->options & NFS_OPTION_MIGRATION)
 		seq_puts(m, ",migration");
@@ -548,9 +543,6 @@
 	else
 		seq_puts(m, ",local_lock=posix");
 
-	if (nfss->flags & NFS_MOUNT_NO_ALIGNWRITE)
-		seq_puts(m, ",noalignwrite");
-
 	if (nfss->flags & NFS_MOUNT_WRITE_EAGER) {
 		if (nfss->flags & NFS_MOUNT_WRITE_WAIT)
 			seq_puts(m, ",write=wait");
@@ -572,6 +564,7 @@
 	seq_printf(m, ",addr=%s",
 			rpc_peeraddr2str(nfss->nfs_client->cl_rpcclient,
 							RPC_DISPLAY_ADDR));
+	nfs_multipath_show_client_info(m, nfss);
 	rcu_read_unlock();
 
 	return 0;
@@ -589,13 +582,18 @@
 	seq_printf(m, ",lease_expired=%ld",
 		   time_after(expire, jiffies) ?  0 : (jiffies - expire) / HZ);
 }
-
+#ifdef CONFIG_NFS_V4_1
 static void show_sessions(struct seq_file *m, struct nfs_server *server)
 {
 	if (nfs4_has_session(server->nfs_client))
 		seq_puts(m, ",sessions");
 }
+#else
+static void show_sessions(struct seq_file *m, struct nfs_server *server) {}
+#endif
+#endif
 
+#ifdef CONFIG_NFS_V4_1
 static void show_pnfs(struct seq_file *m, struct nfs_server *server)
 {
 	seq_printf(m, ",pnfs=");
@@ -615,11 +613,16 @@
 			   impl_id->date.seconds, impl_id->date.nseconds);
 	}
 }
-#else /* CONFIG_NFS_V4 */
+#else
+#if IS_ENABLED(CONFIG_NFS_V4)
+static void show_pnfs(struct seq_file *m, struct nfs_server *server)
+{
+}
+#endif
 static void show_implementation_id(struct seq_file *m, struct nfs_server *nfss)
 {
 }
-#endif /* CONFIG_NFS_V4 */
+#endif
 
 int nfs_show_devname(struct seq_file *m, struct dentry *root)
 {
@@ -666,6 +669,7 @@
 	seq_puts(m, root->d_sb->s_flags & SB_NOATIME ? ",noatime" : "");
 	seq_puts(m, root->d_sb->s_flags & SB_NODIRATIME ? ",nodiratime" : "");
 	nfs_show_mount_options(m, nfss, 1);
+	nfs_multipath_show_client_info(m, nfss);
 
 	seq_printf(m, "\n\tage:\t%lu", (jiffies - nfss->mount_time) / HZ);
 
@@ -901,16 +905,6 @@
 	rpc_authflavor_t authlist[NFS_MAX_SECFLAVORS];
 	unsigned int authlist_len = ARRAY_SIZE(authlist);
 
-	/* make sure 'nolock'/'lock' override the 'local_lock' mount option */
-	if (ctx->lock_status) {
-		if (ctx->lock_status == NFS_LOCK_NOLOCK) {
-			ctx->flags |= NFS_MOUNT_NONLM;
-			ctx->flags |= (NFS_MOUNT_LOCAL_FLOCK | NFS_MOUNT_LOCAL_FCNTL);
-		} else {
-			ctx->flags &= ~NFS_MOUNT_NONLM;
-			ctx->flags &= ~(NFS_MOUNT_LOCAL_FLOCK | NFS_MOUNT_LOCAL_FCNTL);
-		}
-	}
 	status = nfs_request_mount(fc, ctx->mntfh, authlist, &authlist_len);
 	if (status)
 		return ERR_PTR(status);
@@ -1034,6 +1028,16 @@
 	sync_filesystem(sb);
 
 	/*
+	 * The SB_RDONLY flag has been removed from the superblock during
+	 * mounts to prevent interference between different filesystems.
+	 * Similarly, it is also necessary to ignore the SB_RDONLY flag
+	 * during reconfiguration; otherwise, it may also result in the
+	 * creation of redundant superblocks when mounting a directory with
+	 * different rw and ro flags multiple times.
+	 */
+	fc->sb_flags_mask &= ~SB_RDONLY;
+
+	/*
 	 * Userspace mount programs that send binary options generally send
 	 * them populated with default values. We have no way to know which
 	 * ones were explicitly specified. Fall back to legacy behavior and
@@ -1041,7 +1045,17 @@
 	 */
 	if (ctx->skip_reconfig_option_check)
 		return 0;
-
+#if IS_ENABLED(CONFIG_ENFS)
+	if (ctx->enfs_option) {
+		int error = nfs_remount_iplist(nfss->nfs_client, ctx->enfs_option);
+
+		if (error) {
+			/* release remount option member */
+			enfs_free_mount_options(ctx);
+			return error;
+		}
+	}
+#endif
 	/*
 	 * noac is a special case. It implies -o sync, but that's not
 	 * necessarily reflected in the mtab options. reconfigure_super
@@ -1073,9 +1087,8 @@
 	sb->s_blocksize = 0;
 	sb->s_xattr = server->nfs_client->cl_nfs_mod->xattr;
 	sb->s_op = server->nfs_client->cl_nfs_mod->sops;
-	if (server->bsize)
-		sb->s_blocksize =
-			nfs_block_size(server->bsize, &sb->s_blocksize_bits);
+	if (ctx->bsize)
+		sb->s_blocksize = nfs_block_size(ctx->bsize, &sb->s_blocksize_bits);
 
 	switch (server->nfs_client->rpc_ops->version) {
 	case 2:
@@ -1095,7 +1108,7 @@
 		sb->s_export_op = &nfs_export_ops;
 		break;
 	case 4:
-		sb->s_iflags |= SB_I_NOUMASK;
+		sb->s_flags |= SB_POSIXACL;
 		sb->s_time_gran = 1;
 		sb->s_time_min = S64_MIN;
 		sb->s_time_max = S64_MAX;
@@ -1156,7 +1169,7 @@
 	struct nfs_server *server = fc->s_fs_info;
 	int ret;
 
-	set_default_d_op(s, server->nfs_client->rpc_ops->dentry_ops);
+	s->s_d_op = server->nfs_client->rpc_ops->dentry_ops;
 	ret = set_anon_super(s, server);
 	if (ret == 0)
 		server->s_dev = s->s_dev;
@@ -1291,13 +1304,26 @@
 	if (IS_ERR(server))
 		return PTR_ERR(server);
 
+	/*
+	 * When NFS_MOUNT_UNSHARED is not set, NFS forces the sharing of a
+	 * superblock among each filesystem that mounts sub-directories
+	 * belonging to a single exported root path.
+	 * To prevent interference between different filesystems, the
+	 * SB_RDONLY flag should be removed from the superblock.
+	 */
 	if (server->flags & NFS_MOUNT_UNSHARED)
 		compare_super = NULL;
+	else
+		fc->sb_flags &= ~SB_RDONLY;
 
 	/* -o noac implies -o sync */
 	if (server->flags & NFS_MOUNT_NOAC)
 		fc->sb_flags |= SB_SYNCHRONOUS;
 
+	if (ctx->clone_data.sb)
+		if (ctx->clone_data.sb->s_flags & SB_SYNCHRONOUS)
+			fc->sb_flags |= SB_SYNCHRONOUS;
+
 	/* Get a superblock - note that we may end up sharing one that already exists */
 	fc->s_fs_info = server;
 	s = sget_fc(fc, compare_super, nfs_set_super);
@@ -1321,8 +1347,13 @@
 	}
 
 	if (!s->s_root) {
+		unsigned bsize = ctx->clone_data.inherited_bsize;
 		/* initial superblock/root creation */
 		nfs_fill_super(s, ctx);
+		if (bsize) {
+			s->s_blocksize_bits = bsize;
+			s->s_blocksize = 1U << bsize;
+		}
 		error = nfs_get_cache_cookie(s, ctx);
 		if (error < 0)
 			goto error_splat_super;
@@ -1337,6 +1368,10 @@
 	s->s_flags |= SB_ACTIVE;
 	error = 0;
 
+#if IS_ENABLED(CONFIG_ENFS)
+	if (server)
+		enfs_trigger_get_server_capability(server);
+#endif
 out:
 	return error;
 
@@ -1381,7 +1416,6 @@
 unsigned short send_implementation_id = 1;
 char nfs4_client_id_uniquifier[NFS4_CLIENT_ID_UNIQ_LEN] = "";
 bool recover_lost_locks = false;
-short nfs_delay_retrans = -1;
 
 EXPORT_SYMBOL_GPL(nfs_callback_nr_threads);
 EXPORT_SYMBOL_GPL(nfs_callback_set_tcpport);
@@ -1392,7 +1426,6 @@
 EXPORT_SYMBOL_GPL(send_implementation_id);
 EXPORT_SYMBOL_GPL(nfs4_client_id_uniquifier);
 EXPORT_SYMBOL_GPL(recover_lost_locks);
-EXPORT_SYMBOL_GPL(nfs_delay_retrans);
 
 #define NFS_CALLBACK_MAXPORTNR (65535U)
 
@@ -1441,9 +1474,5 @@
 		 "If the server reports that a lock might be lost, "
 		 "try to recover it risking data corruption.");
 
-module_param_named(delay_retrans, nfs_delay_retrans, short, 0644);
-MODULE_PARM_DESC(delay_retrans,
-		 "Unless negative, specifies the number of times the NFSv4 "
-		 "client retries a request before returning an EAGAIN error, "
-		 "after a reply of NFS4ERR_DELAY from the server.");
+
 #endif /* CONFIG_NFS_V4 */
