--- /tmp/tmp.tYSqORsSZN/ubuntu	2026-04-30 18:36:47.670061753 +0200
+++ /home/darren/development/enfs/vendor/openeuler/fs/nfs/internal.h	2026-04-30 17:42:52.000000000 +0200
@@ -6,11 +6,9 @@
 #include "nfs4_fs.h"
 #include <linux/fs_context.h>
 #include <linux/security.h>
-#include <linux/compiler_attributes.h>
 #include <linux/crc32.h>
 #include <linux/sunrpc/addr.h>
 #include <linux/nfs_page.h>
-#include <linux/nfslocalio.h>
 #include <linux/wait_bit.h>
 
 #define NFS_SB_MASK (SB_RDONLY|SB_NOSUID|SB_NODEV|SB_NOEXEC|SB_SYNCHRONOUS)
@@ -86,6 +84,9 @@
 	struct xprtsec_parms xprtsec;
 	unsigned long connect_timeout;
 	unsigned long reconnect_timeout;
+#if IS_ENABLED(CONFIG_ENFS)
+	void *enfs_option; /* struct multipath_mount_options */
+#endif
 };
 
 /*
@@ -114,7 +115,6 @@
 	unsigned short		protofamily;
 	unsigned short		mountfamily;
 	bool			has_sec_mnt_opts;
-	int			lock_status;
 
 	struct {
 		union {
@@ -152,13 +152,11 @@
 		struct super_block	*sb;
 		struct dentry		*dentry;
 		struct nfs_fattr	*fattr;
+		unsigned int		inherited_bsize;
 	} clone_data;
-};
-
-enum nfs_lock_status {
-	NFS_LOCK_NOT_SET	= 0,
-	NFS_LOCK_LOCK		= 1,
-	NFS_LOCK_NOLOCK		= 2,
+#if IS_ENABLED(CONFIG_ENFS)
+	void *enfs_option; /* struct multipath_mount_options */
+#endif
 };
 
 #define nfs_errorf(fc, fmt, ...) ((fc)->log.log ?		\
@@ -206,6 +204,7 @@
 };
 
 extern int nfs_mount(struct nfs_mount_request *info, int timeo, int retrans);
+extern void nfs_umount(const struct nfs_mount_request *info);
 
 /* client.c */
 extern const struct rpc_program nfs_program;
@@ -230,7 +229,7 @@
 nfs4_find_client_sessionid(struct net *, const struct sockaddr *,
 				struct nfs4_sessionid *, u32);
 extern struct nfs_server *nfs_create_server(struct fs_context *);
-extern void nfs_server_set_init_caps(struct nfs_server *);
+extern void nfs4_server_set_init_caps(struct nfs_server *);
 extern struct nfs_server *nfs4_create_server(struct fs_context *);
 extern struct nfs_server *nfs4_create_referral_server(struct fs_context *);
 extern int nfs4_update_server(struct nfs_server *server, const char *hostname,
@@ -308,8 +307,7 @@
 int nfs_generic_pgio(struct nfs_pageio_descriptor *, struct nfs_pgio_header *);
 int nfs_initiate_pgio(struct rpc_clnt *clnt, struct nfs_pgio_header *hdr,
 		      const struct cred *cred, const struct nfs_rpc_ops *rpc_ops,
-		      const struct rpc_call_ops *call_ops, int how, int flags,
-		      struct nfsd_file *localio);
+		      const struct rpc_call_ops *call_ops, int how, int flags);
 void nfs_free_request(struct nfs_page *req);
 struct nfs_pgio_mirror *
 nfs_pgio_current_mirror(struct nfs_pageio_descriptor *desc);
@@ -334,13 +332,17 @@
 #if IS_ENABLED(CONFIG_NFS_V4)
 extern int nfs4_decode_dirent(struct xdr_stream *,
 				struct nfs_entry *, bool);
+#endif
+#ifdef CONFIG_NFS_V4_1
 extern const u32 nfs41_maxread_overhead;
 extern const u32 nfs41_maxwrite_overhead;
 extern const u32 nfs41_maxgetdevinfo_overhead;
+#endif
 
 /* nfs4proc.c */
+#if IS_ENABLED(CONFIG_NFS_V4)
 extern const struct rpc_procinfo nfs4_procedures[];
-#endif /* CONFIG_NFS_V4 */
+#endif
 
 #ifdef CONFIG_NFS_V4_SECURITY_LABEL
 extern struct nfs4_label *nfs4_label_alloc(struct nfs_server *server, gfp_t flags);
@@ -386,16 +388,16 @@
 /* dir.c */
 extern void nfs_readdir_record_entry_cache_hit(struct inode *dir);
 extern void nfs_readdir_record_entry_cache_miss(struct inode *dir);
-extern unsigned long nfs_access_cache_count(struct shrinker *shrink,
+extern unsigned long nfs_access_cache_count(struct shrinker_v2 *shrink,
 					    struct shrink_control *sc);
-extern unsigned long nfs_access_cache_scan(struct shrinker *shrink,
+extern unsigned long nfs_access_cache_scan(struct shrinker_v2 *shrink,
 					   struct shrink_control *sc);
 struct dentry *nfs_lookup(struct inode *, struct dentry *, unsigned int);
 void nfs_d_prune_case_insensitive_aliases(struct inode *inode);
 int nfs_create(struct mnt_idmap *, struct inode *, struct dentry *,
 	       umode_t, bool);
-struct dentry *nfs_mkdir(struct mnt_idmap *, struct inode *, struct dentry *,
-			 umode_t);
+int nfs_mkdir(struct mnt_idmap *, struct inode *, struct dentry *,
+	      umode_t);
 int nfs_rmdir(struct inode *, struct dentry *);
 int nfs_unlink(struct inode *, struct dentry *);
 int nfs_symlink(struct mnt_idmap *, struct inode *, struct dentry *,
@@ -426,18 +428,15 @@
 ssize_t nfs_file_read(struct kiocb *, struct iov_iter *);
 ssize_t nfs_file_splice_read(struct file *in, loff_t *ppos, struct pipe_inode_info *pipe,
 			     size_t len, unsigned int flags);
-int nfs_file_mmap_prepare(struct vm_area_desc *);
+int nfs_file_mmap(struct file *, struct vm_area_struct *);
 ssize_t nfs_file_write(struct kiocb *, struct iov_iter *);
 int nfs_file_release(struct inode *, struct file *);
 int nfs_lock(struct file *, int, struct file_lock *);
 int nfs_flock(struct file *, int, struct file_lock *);
 int nfs_check_flags(int);
-void nfs_truncate_last_folio(struct address_space *mapping, loff_t from,
-			     loff_t to);
 
 /* inode.c */
 extern struct workqueue_struct *nfsiod_workqueue;
-extern struct workqueue_struct *nfslocaliod_workqueue;
 extern struct inode *nfs_alloc_inode(struct super_block *sb);
 extern void nfs_free_inode(struct inode *);
 extern int nfs_write_inode(struct inode *, struct writeback_control *);
@@ -449,63 +448,6 @@
 extern bool nfs_check_cache_invalid(struct inode *, unsigned long);
 extern int nfs_wait_bit_killable(struct wait_bit_key *key, int mode);
 
-#if IS_ENABLED(CONFIG_NFS_LOCALIO)
-/* localio.c */
-struct nfs_local_dio {
-	u32 mem_align;
-	u32 offset_align;
-	loff_t middle_offset;
-	loff_t end_offset;
-	ssize_t	start_len;	/* Length for misaligned first extent */
-	ssize_t	middle_len;	/* Length for DIO-aligned middle extent */
-	ssize_t	end_len;	/* Length for misaligned last extent */
-};
-
-extern void nfs_local_probe_async(struct nfs_client *);
-extern void nfs_local_probe_async_work(struct work_struct *);
-extern struct nfsd_file *nfs_local_open_fh(struct nfs_client *,
-					   const struct cred *,
-					   struct nfs_fh *,
-					   struct nfs_file_localio *,
-					   const fmode_t);
-extern int nfs_local_doio(struct nfs_client *,
-			  struct nfsd_file *,
-			  struct nfs_pgio_header *,
-			  const struct rpc_call_ops *);
-extern int nfs_local_commit(struct nfsd_file *,
-			    struct nfs_commit_data *,
-			    const struct rpc_call_ops *, int);
-extern bool nfs_server_is_local(const struct nfs_client *clp);
-
-#else /* CONFIG_NFS_LOCALIO */
-static inline void nfs_local_probe(struct nfs_client *clp) {}
-static inline void nfs_local_probe_async(struct nfs_client *clp) {}
-static inline struct nfsd_file *
-nfs_local_open_fh(struct nfs_client *clp, const struct cred *cred,
-		  struct nfs_fh *fh, struct nfs_file_localio *nfl,
-		  const fmode_t mode)
-{
-	return NULL;
-}
-static inline int nfs_local_doio(struct nfs_client *clp,
-				 struct nfsd_file *localio,
-				 struct nfs_pgio_header *hdr,
-				 const struct rpc_call_ops *call_ops)
-{
-	return -EINVAL;
-}
-static inline int nfs_local_commit(struct nfsd_file *localio,
-				struct nfs_commit_data *data,
-				const struct rpc_call_ops *call_ops, int how)
-{
-	return -EINVAL;
-}
-static inline bool nfs_server_is_local(const struct nfs_client *clp)
-{
-	return false;
-}
-#endif /* CONFIG_NFS_LOCALIO */
-
 /* super.c */
 extern const struct super_operations nfs_sops;
 bool nfs_auth_info_match(const struct nfs_auth_info *, rpc_authflavor_t);
@@ -525,11 +467,11 @@
 #endif
 
 /* io.c */
-extern __must_check int nfs_start_io_read(struct inode *inode);
+extern void nfs_start_io_read(struct inode *inode);
 extern void nfs_end_io_read(struct inode *inode);
-extern  __must_check int nfs_start_io_write(struct inode *inode);
+extern void nfs_start_io_write(struct inode *inode);
 extern void nfs_end_io_write(struct inode *inode);
-extern __must_check int nfs_start_io_direct(struct inode *inode);
+extern void nfs_start_io_direct(struct inode *inode);
 extern void nfs_end_io_direct(struct inode *inode);
 
 static inline bool nfs_file_io_is_buffered(struct nfs_inode *nfsi)
@@ -537,16 +479,6 @@
 	return test_bit(NFS_INO_ODIRECT, &nfsi->flags) == 0;
 }
 
-/* Must be called with exclusively locked inode->i_rwsem */
-static inline void nfs_file_block_o_direct(struct nfs_inode *nfsi)
-{
-	if (test_bit(NFS_INO_ODIRECT, &nfsi->flags)) {
-		clear_bit(NFS_INO_ODIRECT, &nfsi->flags);
-		inode_dio_wait(&nfsi->vfs_inode);
-	}
-}
-
-
 /* namespace.c */
 #define NFS_PATH_CANONICAL 1
 extern char *nfs_path(char **p, struct dentry *dentry,
@@ -572,6 +504,7 @@
 			       struct nfs_open_context *ctx,
 			       struct folio *folio);
 extern void nfs_pageio_complete_read(struct nfs_pageio_descriptor *pgio);
+extern void nfs_read_prepare(struct rpc_task *task, void *calldata);
 extern void nfs_pageio_reset_read_mds(struct nfs_pageio_descriptor *pgio);
 
 /* super.c */
@@ -594,8 +527,7 @@
 			       struct nfs_commit_data *data,
 			       const struct nfs_rpc_ops *nfs_ops,
 			       const struct rpc_call_ops *call_ops,
-			       int how, int flags,
-			       struct nfsd_file *localio);
+			       int how, int flags);
 extern void nfs_init_commit(struct nfs_commit_data *data,
 			    struct list_head *head,
 			    struct pnfs_layout_segment *lseg,
@@ -635,7 +567,7 @@
 int nfs_filemap_write_and_wait_range(struct address_space *mapping,
 		loff_t lstart, loff_t lend);
 
-#ifdef CONFIG_NFS_V4
+#ifdef CONFIG_NFS_V4_1
 static inline void
 pnfs_bucket_clear_pnfs_ds_commit_verifiers(struct pnfs_commit_bucket *buckets,
 		unsigned int nbuckets)
@@ -656,12 +588,12 @@
 				array->nbuckets);
 	rcu_read_unlock();
 }
-#else /* CONFIG_NFS_V4 */
+#else
 static inline
 void nfs_clear_pnfs_ds_commit_verifiers(struct pnfs_ds_commit_info *cinfo)
 {
 }
-#endif /* CONFIG_NFS_V4 */
+#endif
 
 #ifdef CONFIG_MIGRATION
 int nfs_migrate_folio(struct address_space *, struct folio *dst,
@@ -687,12 +619,9 @@
 
 static inline gfp_t nfs_io_gfp_mask(void)
 {
-	gfp_t ret = current_gfp_context(GFP_KERNEL);
-
-	/* For workers __GFP_NORETRY only with __GFP_IO or __GFP_FS */
-	if ((current->flags & PF_WQ_WORKER) && ret == GFP_KERNEL)
-		ret |= __GFP_NORETRY | __GFP_NOWARN;
-	return ret;
+	if (current->flags & PF_WQ_WORKER)
+		return GFP_KERNEL | __GFP_NORETRY | __GFP_NOWARN;
+	return GFP_KERNEL;
 }
 
 /*
@@ -735,6 +664,9 @@
 /* nfs4proc.c */
 extern struct nfs_client *nfs4_init_client(struct nfs_client *clp,
 			    const struct nfs_client_initdata *);
+extern int nfs40_walk_client_list(struct nfs_client *clp,
+				struct nfs_client **result,
+				const struct cred *cred);
 extern int nfs41_walk_client_list(struct nfs_client *clp,
 				struct nfs_client **result,
 				const struct cred *cred);
@@ -852,7 +784,7 @@
 					   struct nfs_commit_info *cinfo)
 {
 	if (folio && !cinfo->dreq) {
-		struct inode *inode = folio->mapping->host;
+		struct inode *inode = folio_file_mapping(folio)->host;
 		long nr = folio_nr_pages(folio);
 
 		/* This page is really still in write-back - just that the
@@ -867,12 +799,31 @@
 /*
  * Determine the number of bytes of data the page contains
  */
+static inline
+unsigned int nfs_page_length(struct page *page)
+{
+	loff_t i_size = i_size_read(page_file_mapping(page)->host);
+
+	if (i_size > 0) {
+		pgoff_t index = page_index(page);
+		pgoff_t end_index = (i_size - 1) >> PAGE_SHIFT;
+		if (index < end_index)
+			return PAGE_SIZE;
+		if (index == end_index)
+			return ((i_size - 1) & ~PAGE_MASK) + 1;
+	}
+	return 0;
+}
+
+/*
+ * Determine the number of bytes of data the page contains
+ */
 static inline size_t nfs_folio_length(struct folio *folio)
 {
-	loff_t i_size = i_size_read(folio->mapping->host);
+	loff_t i_size = i_size_read(folio_file_mapping(folio)->host);
 
 	if (i_size > 0) {
-		pgoff_t index = folio->index >> folio_order(folio);
+		pgoff_t index = folio_index(folio) >> folio_order(folio);
 		pgoff_t end_index = (i_size - 1) >> folio_shift(folio);
 		if (index < end_index)
 			return folio_size(folio);
@@ -920,11 +871,6 @@
 				NFS4_STATEID_OTHER_SIZE);
 }
 
-static inline bool nfs_current_task_exiting(void)
-{
-	return (current->flags & PF_EXITING) != 0;
-}
-
 static inline bool nfs_error_is_fatal(int err)
 {
 	switch (err) {
@@ -987,6 +933,7 @@
 	loff_t			io_start;	/* Start offset for I/O */
 	ssize_t			count,		/* bytes actually processed */
 				max_count,	/* max expected count */
+				bytes_left,	/* bytes left to be sent */
 				error;		/* any reported error */
 	struct completion	completion;	/* wait for i/o completion */
 
