--- /tmp/tmp.tYSqORsSZN/ubuntu	2026-04-30 18:36:45.554062129 +0200
+++ /home/darren/development/enfs/vendor/openeuler/fs/nfs/fs_context.c	2026-04-30 17:42:52.000000000 +0200
@@ -22,7 +22,7 @@
 #include <net/handshake.h>
 
 #include "nfs.h"
-#include "internal.h"
+#include "enfs_adapter.h"
 
 #include "nfstrace.h"
 
@@ -49,8 +49,6 @@
 	Opt_bsize,
 	Opt_clientaddr,
 	Opt_cto,
-	Opt_alignwrite,
-	Opt_fatal_neterrors,
 	Opt_fg,
 	Opt_fscache,
 	Opt_fscache_flag,
@@ -73,8 +71,6 @@
 	Opt_posix,
 	Opt_proto,
 	Opt_rdirplus,
-	Opt_rdirplus_none,
-	Opt_rdirplus_force,
 	Opt_rdma,
 	Opt_resvport,
 	Opt_retrans,
@@ -96,22 +92,13 @@
 	Opt_wsize,
 	Opt_write,
 	Opt_xprtsec,
-	Opt_cert_serial,
-	Opt_privkey_serial,
-};
-
-enum {
-	Opt_fatal_neterrors_default,
-	Opt_fatal_neterrors_enetunreach,
-	Opt_fatal_neterrors_none,
-};
-
-static const struct constant_table nfs_param_enums_fatal_neterrors[] = {
-	{ "default",			Opt_fatal_neterrors_default },
-	{ "ENETDOWN:ENETUNREACH",	Opt_fatal_neterrors_enetunreach },
-	{ "ENETUNREACH:ENETDOWN",	Opt_fatal_neterrors_enetunreach },
-	{ "none",			Opt_fatal_neterrors_none },
-	{}
+#if IS_ENABLED(CONFIG_ENFS)
+	Opt_remote_addrs,
+	Opt_local_iplist,
+	Opt_enfs_info,
+	Opt_slookupcache,
+	Opt_alookupcache,
+#endif
 };
 
 enum {
@@ -169,9 +156,6 @@
 	fsparam_u32   ("bsize",		Opt_bsize),
 	fsparam_string("clientaddr",	Opt_clientaddr),
 	fsparam_flag_no("cto",		Opt_cto),
-	fsparam_flag_no("alignwrite",	Opt_alignwrite),
-	fsparam_enum("fatal_neterrors", Opt_fatal_neterrors,
-		     nfs_param_enums_fatal_neterrors),
 	fsparam_flag  ("fg",		Opt_fg),
 	fsparam_flag_no("fsc",		Opt_fscache_flag),
 	fsparam_string("fsc",		Opt_fscache),
@@ -195,8 +179,7 @@
 	fsparam_u32   ("port",		Opt_port),
 	fsparam_flag_no("posix",	Opt_posix),
 	fsparam_string("proto",		Opt_proto),
-	fsparam_flag_no("rdirplus", Opt_rdirplus), // rdirplus|nordirplus
-	fsparam_string("rdirplus",  Opt_rdirplus), // rdirplus=...
+	fsparam_flag_no("rdirplus",	Opt_rdirplus),
 	fsparam_flag  ("rdma",		Opt_rdma),
 	fsparam_flag_no("resvport",	Opt_resvport),
 	fsparam_u32   ("retrans",	Opt_retrans),
@@ -223,8 +206,13 @@
 	fsparam_enum  ("write",		Opt_write, nfs_param_enums_write),
 	fsparam_u32   ("wsize",		Opt_wsize),
 	fsparam_string("xprtsec",	Opt_xprtsec),
-	fsparam_s32("cert_serial",	Opt_cert_serial),
-	fsparam_s32("privkey_serial",	Opt_privkey_serial),
+#if IS_ENABLED(CONFIG_ENFS)
+	fsparam_string("localaddrs",	Opt_local_iplist),
+	fsparam_string("remoteaddrs",	Opt_remote_addrs),
+	fsparam_string("enfs_info",	Opt_enfs_info),
+	fsparam_string("slookupcache",	Opt_slookupcache),
+	fsparam_string("alookupcache",	Opt_alookupcache),
+#endif
 	{}
 };
 
@@ -312,12 +300,6 @@
 	{}
 };
 
-static const struct constant_table nfs_rdirplus_tokens[] = {
-	{ "none",	Opt_rdirplus_none },
-	{ "force",	Opt_rdirplus_force },
-	{}
-};
-
 /*
  * Sanity-check a server address provided by the mount command.
  *
@@ -388,6 +370,20 @@
 	return nfs_invalf(fc, "NFS: Transport does not support xprtsec");
 }
 
+#if IS_ENABLED(CONFIG_ENFS)
+enum nfsmultipathoptions getNfsMultiPathOpt(int token)
+{
+	switch (token) {
+	case Opt_remote_addrs: {
+		return REMOTEADDR;
+	}
+	case Opt_local_iplist: {
+		return LOCALADDR;
+	}
+	}
+	return INVALID_OPTION;
+}
+#endif
 /*
  * For text based NFSv2/v3 mounts, the mount protocol transport default
  * settings should depend upon the specified NFS transport.
@@ -555,32 +551,6 @@
 	return 0;
 }
 
-#ifdef CONFIG_KEYS
-static int nfs_tls_key_verify(key_serial_t key_id)
-{
-	struct key *key = key_lookup(key_id);
-	int error = 0;
-
-	if (IS_ERR(key)) {
-		pr_err("key id %08x not found\n", key_id);
-		return PTR_ERR(key);
-	}
-	if (test_bit(KEY_FLAG_REVOKED, &key->flags) ||
-	    test_bit(KEY_FLAG_INVALIDATED, &key->flags)) {
-		pr_err("key id %08x revoked\n", key_id);
-		error = -EKEYREVOKED;
-	}
-
-	key_put(key);
-	return error;
-}
-#else
-static inline int nfs_tls_key_verify(key_serial_t key_id)
-{
-	return -ENOENT;
-}
-#endif /* CONFIG_KEYS */
-
 /*
  * Parse a single mount parameter.
  */
@@ -650,12 +620,6 @@
 		else
 			ctx->flags |= NFS_MOUNT_TRUNK_DISCOVERY;
 		break;
-	case Opt_alignwrite:
-		if (result.negated)
-			ctx->flags |= NFS_MOUNT_NO_ALIGNWRITE;
-		else
-			ctx->flags &= ~NFS_MOUNT_NO_ALIGNWRITE;
-		break;
 	case Opt_ac:
 		if (result.negated)
 			ctx->flags |= NFS_MOUNT_NOAC;
@@ -664,11 +628,9 @@
 		break;
 	case Opt_lock:
 		if (result.negated) {
-			ctx->lock_status = NFS_LOCK_NOLOCK;
 			ctx->flags |= NFS_MOUNT_NONLM;
 			ctx->flags |= (NFS_MOUNT_LOCAL_FLOCK | NFS_MOUNT_LOCAL_FCNTL);
 		} else {
-			ctx->lock_status = NFS_LOCK_LOCK;
 			ctx->flags &= ~NFS_MOUNT_NONLM;
 			ctx->flags &= ~(NFS_MOUNT_LOCAL_FLOCK | NFS_MOUNT_LOCAL_FCNTL);
 		}
@@ -692,25 +654,10 @@
 			ctx->flags &= ~NFS_MOUNT_NOACL;
 		break;
 	case Opt_rdirplus:
-		if (result.negated) {
-			ctx->flags &= ~NFS_MOUNT_FORCE_RDIRPLUS;
+		if (result.negated)
 			ctx->flags |= NFS_MOUNT_NORDIRPLUS;
-		} else if (!param->string) {
-			ctx->flags &= ~(NFS_MOUNT_NORDIRPLUS | NFS_MOUNT_FORCE_RDIRPLUS);
-		} else {
-			switch (lookup_constant(nfs_rdirplus_tokens, param->string, -1)) {
-			case Opt_rdirplus_none:
-				ctx->flags &= ~NFS_MOUNT_FORCE_RDIRPLUS;
-				ctx->flags |= NFS_MOUNT_NORDIRPLUS;
-				break;
-			case Opt_rdirplus_force:
-				ctx->flags &= ~NFS_MOUNT_NORDIRPLUS;
-				ctx->flags |= NFS_MOUNT_FORCE_RDIRPLUS;
-				break;
-			default:
-				goto out_invalid_value;
-			}
-		}
+		else
+			ctx->flags &= ~NFS_MOUNT_NORDIRPLUS;
 		break;
 	case Opt_sharecache:
 		if (result.negated)
@@ -733,7 +680,6 @@
 		ctx->fscache_uniq = NULL;
 		break;
 	case Opt_fscache:
-		trace_nfs_mount_assign(param->key, param->string);
 		ctx->options |= NFS_OPTION_FSCACHE;
 		kfree(ctx->fscache_uniq);
 		ctx->fscache_uniq = param->string;
@@ -806,8 +752,7 @@
 		ctx->mount_server.version = result.uint_32;
 		break;
 	case Opt_minorversion:
-		if (result.uint_32 < NFS4_MIN_MINOR_VERSION ||
-		    result.uint_32 > NFS4_MAX_MINOR_VERSION)
+		if (result.uint_32 > NFS4_MAX_MINOR_VERSION)
 			goto out_of_bounds;
 		ctx->minorversion = result.uint_32;
 		break;
@@ -838,18 +783,6 @@
 		if (ret < 0)
 			return ret;
 		break;
-	case Opt_cert_serial:
-		ret = nfs_tls_key_verify(result.int_32);
-		if (ret < 0)
-			return ret;
-		ctx->xprtsec.cert_serial = result.int_32;
-		break;
-	case Opt_privkey_serial:
-		ret = nfs_tls_key_verify(result.int_32);
-		if (ret < 0)
-			return ret;
-		ctx->xprtsec.privkey_serial = result.int_32;
-		break;
 
 	case Opt_proto:
 		if (!param->string)
@@ -956,25 +889,6 @@
 			goto out_of_bounds;
 		ctx->nfs_server.max_connect = result.uint_32;
 		break;
-	case Opt_fatal_neterrors:
-		trace_nfs_mount_assign(param->key, param->string);
-		switch (result.uint_32) {
-		case Opt_fatal_neterrors_default:
-			if (fc->net_ns != &init_net)
-				ctx->flags |= NFS_MOUNT_NETUNREACH_FATAL;
-			else
-				ctx->flags &= ~NFS_MOUNT_NETUNREACH_FATAL;
-			break;
-		case Opt_fatal_neterrors_enetunreach:
-			ctx->flags |= NFS_MOUNT_NETUNREACH_FATAL;
-			break;
-		case Opt_fatal_neterrors_none:
-			ctx->flags &= ~NFS_MOUNT_NETUNREACH_FATAL;
-			break;
-		default:
-			goto out_invalid_value;
-		}
-		break;
 	case Opt_lookupcache:
 		trace_nfs_mount_assign(param->key, param->string);
 		switch (result.uint_32) {
@@ -1039,6 +953,28 @@
 	case Opt_sloppy:
 		ctx->sloppy = true;
 		break;
+#if IS_ENABLED(CONFIG_ENFS)
+	case Opt_local_iplist:
+	case Opt_remote_addrs:
+		switch (enfs_parse_mount_options(getNfsMultiPathOpt(opt),
+						 param->string, ctx, fc)) {
+		case 0:
+			break;
+		case -ENOMEM:
+			goto out_nomem;
+		case -ENOSPC:
+			goto out_limit;
+		case -EINVAL:
+			goto out_invalid_address;
+		case -EOPNOTSUPP:
+			goto out_invalid_address;
+		}
+		break;
+	case Opt_enfs_info:
+	case Opt_slookupcache:
+	case Opt_alookupcache:
+		break;
+#endif
 	}
 
 	return 0;
@@ -1051,6 +987,12 @@
 	return nfs_invalf(fc, "NFS: Value for '%s' out of range", param->key);
 out_bad_transport:
 	return nfs_invalf(fc, "NFS: Unrecognized transport protocol");
+#if IS_ENABLED(CONFIG_ENFS)
+out_limit:
+	return nfs_invalf(fc, "NFS: param is more than supported limit");
+out_nomem:
+	return nfs_invalf(fc, "NFS: not enough memory to parse option");
+#endif
 }
 
 /*
@@ -1270,7 +1212,8 @@
 			int ret;
 
 			data->context[NFS_MAX_CONTEXT_LEN] = '\0';
-			ret = vfs_parse_fs_string(fc, "context", data->context);
+			ret = vfs_parse_fs_string(fc, "context",
+						  data->context, strlen(data->context));
 			if (ret < 0)
 				return ret;
 #else
@@ -1566,10 +1509,12 @@
 	ret = nfs_parse_source(fc, max_namelen, max_pathlen);
 	if (ret < 0)
 		return ret;
-
+#if IS_ENABLED(CONFIG_ENFS)
+	nfs_multipath_set_mount_data(&ctx->enfs_option, ctx->nfs_server.hostname);
+#endif
 	/* Load the NFS protocol module if we haven't done so yet */
 	if (!ctx->nfs_mod) {
-		nfs_mod = find_nfs_version(ctx->version);
+		nfs_mod = get_nfs_version(ctx->version);
 		if (IS_ERR(nfs_mod)) {
 			ret = PTR_ERR(nfs_mod);
 			goto out_version_unavailable;
@@ -1643,13 +1588,16 @@
 	}
 	nfs_copy_fh(ctx->mntfh, src->mntfh);
 
-	get_nfs_version(ctx->nfs_mod);
+	__module_get(ctx->nfs_mod->owner);
 	ctx->client_address		= NULL;
 	ctx->mount_server.hostname	= NULL;
 	ctx->nfs_server.export_path	= NULL;
 	ctx->nfs_server.hostname	= NULL;
 	ctx->fscache_uniq		= NULL;
 	ctx->clone_data.fattr		= NULL;
+#if IS_ENABLED(CONFIG_ENFS)
+	ctx->enfs_option = NULL;
+#endif
 	fc->fs_private = ctx;
 	return 0;
 }
@@ -1668,6 +1616,7 @@
 		kfree(ctx->nfs_server.export_path);
 		kfree(ctx->nfs_server.hostname);
 		kfree(ctx->fscache_uniq);
+		enfs_free_mount_options(ctx);
 		nfs_free_fhandle(ctx->mntfh);
 		nfs_free_fattr(ctx->clone_data.fattr);
 		kfree(ctx);
@@ -1692,7 +1641,7 @@
 {
 	struct nfs_fs_context *ctx;
 
-	ctx = kzalloc_obj(struct nfs_fs_context);
+	ctx = kzalloc(sizeof(struct nfs_fs_context), GFP_KERNEL);
 	if (unlikely(!ctx))
 		return -ENOMEM;
 
@@ -1735,7 +1684,7 @@
 		}
 
 		ctx->nfs_mod = nfss->nfs_client->cl_nfs_mod;
-		get_nfs_version(ctx->nfs_mod);
+		__module_get(ctx->nfs_mod->owner);
 	} else {
 		/* defaults */
 		ctx->timeo		= NFS_UNSPEC_TIMEO;
@@ -1753,9 +1702,6 @@
 		ctx->xprtsec.cert_serial	= TLS_NO_CERT;
 		ctx->xprtsec.privkey_serial	= TLS_NO_PRIVKEY;
 
-		if (fc->net_ns != &init_net)
-			ctx->flags |= NFS_MOUNT_NETUNREACH_FATAL;
-
 		fc->s_iflags		|= SB_I_STABLE_WRITES;
 	}
 	fc->fs_private = ctx;
