--- /tmp/tmp.tYSqORsSZN/ubuntu	2026-04-30 18:36:59.382059666 +0200
+++ /home/darren/development/enfs/vendor/openeuler/include/linux/sunrpc/clnt.h	2026-04-30 17:42:52.000000000 +0200
@@ -28,6 +28,7 @@
 #include <linux/path.h>
 #include <net/ipv6.h>
 #include <linux/sunrpc/xprtmultipath.h>
+#include <linux/kabi.h>
 
 struct rpc_inode;
 struct rpc_sysfs_client {
@@ -37,7 +38,6 @@
 	struct rpc_xprt_switch *xprt_switch;
 };
 
-
 /*
  * The high-level client handle
  */
@@ -64,9 +64,12 @@
 				cl_noretranstimeo: 1,/* No retransmit timeouts */
 				cl_autobind : 1,/* use getport() */
 				cl_chatty   : 1,/* be verbose */
+#if defined(__GENKSYMS__) || !IS_ENABLED(CONFIG_SUNRPC_ENFS)
+				cl_shutdown : 1;/* rpc immediate -EIO */
+#else
 				cl_shutdown : 1,/* rpc immediate -EIO */
-				cl_netunreach_fatal : 1;
-						/* Treat ENETUNREACH errors as fatal */
+				cl_enfs     : 1;/* be enfs */
+#endif
 	struct xprtsec_parms	cl_xprtsec;	/* transport security policy */
 
 	struct rpc_rtt *	cl_rtt;		/* RTO estimator data */
@@ -95,7 +98,11 @@
 	const struct cred	*cl_cred;
 	unsigned int		cl_max_connect; /* max number of transports not to the same IP */
 	struct super_block *pipefs_sb;
-	atomic_t		cl_task_count;
+
+	KABI_RESERVE(1)
+	KABI_RESERVE(2)
+	KABI_RESERVE(3)
+	KABI_RESERVE(4)
 };
 
 /*
@@ -109,6 +116,9 @@
 	const struct rpc_version **	version;	/* version array */
 	struct rpc_stat *	stats;		/* statistics */
 	const char *		pipe_dir_name;	/* path to rpc_pipefs dir */
+
+	KABI_RESERVE(1)
+	KABI_RESERVE(2)
 };
 
 struct rpc_version {
@@ -155,6 +165,13 @@
 	struct xprtsec_parms	xprtsec;
 	unsigned long		connect_timeout;
 	unsigned long		reconnect_timeout;
+
+#if IS_ENABLED(CONFIG_SUNRPC_ENFS)
+	void *multipath_option;
+#else
+	KABI_RESERVE(1)
+#endif
+	KABI_RESERVE(2)
 };
 
 struct rpc_add_xprt_test {
@@ -177,7 +194,6 @@
 #define RPC_CLNT_CREATE_SOFTERR		(1UL << 10)
 #define RPC_CLNT_CREATE_REUSEPORT	(1UL << 11)
 #define RPC_CLNT_CREATE_CONNECTED	(1UL << 12)
-#define RPC_CLNT_CREATE_NETUNREACH_FATAL	(1UL << 13)
 
 struct rpc_clnt *rpc_create(struct rpc_create_args *args);
 struct rpc_clnt	*rpc_bind_new_program(struct rpc_clnt *,
@@ -228,6 +244,7 @@
 size_t		rpc_peeraddr(struct rpc_clnt *, struct sockaddr *, size_t);
 const char	*rpc_peeraddr2str(struct rpc_clnt *, enum rpc_display_format_t);
 int		rpc_localaddr(struct rpc_clnt *, struct sockaddr *, size_t);
+int rpc_localalladdr(struct rpc_xprt *xprt, struct sockaddr *buf, size_t buflen);
 
 int 		rpc_clnt_iterate_for_each_xprt(struct rpc_clnt *clnt,
 			int (*fn)(struct rpc_clnt *, struct rpc_xprt *, void *),
@@ -257,6 +274,7 @@
 
 const char *rpc_proc_name(const struct rpc_task *task);
 
+void rpc_clnt_xprt_switch_put(struct rpc_clnt *);
 void rpc_clnt_xprt_switch_add_xprt(struct rpc_clnt *, struct rpc_xprt *);
 void rpc_clnt_xprt_switch_remove_xprt(struct rpc_clnt *, struct rpc_xprt *);
 bool rpc_clnt_xprt_switch_has_addr(struct rpc_clnt *clnt,
@@ -264,6 +282,10 @@
 void rpc_clnt_xprt_set_online(struct rpc_clnt *clnt, struct rpc_xprt *xprt);
 void rpc_clnt_disconnect(struct rpc_clnt *clnt);
 void rpc_cleanup_clids(void);
+int rpc_clnt_test_xprt(struct rpc_clnt *clnt, struct rpc_xprt *xprt,
+	const struct rpc_call_ops *ops, void *data, int flags);
+
+struct rpc_xprt *rpc_task_get_next_xprt(struct rpc_clnt *clnt);
 
 static inline int rpc_reply_expected(struct rpc_task *task)
 {
