--- /tmp/tmp.tYSqORsSZN/ubuntu	2026-04-30 18:36:56.346060198 +0200
+++ /home/darren/development/enfs/vendor/openeuler/include/linux/nfs_xdr.h	2026-04-30 17:42:52.000000000 +0200
@@ -45,7 +45,7 @@
 };
 
 struct nfs_fattr {
-	__u64			valid;		/* which fields are valid */
+	unsigned int		valid;		/* which fields are valid */
 	umode_t			mode;
 	__u32			nlink;
 	kuid_t			uid;
@@ -67,7 +67,6 @@
 	struct timespec64	atime;
 	struct timespec64	mtime;
 	struct timespec64	ctime;
-	struct timespec64	btime;
 	__u64			change_attr;	/* NFSv4 change attribute */
 	__u64			pre_change_attr;/* pre-op NFSv4 change attribute */
 	__u64			pre_size;	/* pre_op_attr.size	  */
@@ -81,33 +80,32 @@
 	struct nfs4_label	*label;
 };
 
-#define NFS_ATTR_FATTR_TYPE		BIT_ULL(0)
-#define NFS_ATTR_FATTR_MODE		BIT_ULL(1)
-#define NFS_ATTR_FATTR_NLINK		BIT_ULL(2)
-#define NFS_ATTR_FATTR_OWNER		BIT_ULL(3)
-#define NFS_ATTR_FATTR_GROUP		BIT_ULL(4)
-#define NFS_ATTR_FATTR_RDEV		BIT_ULL(5)
-#define NFS_ATTR_FATTR_SIZE		BIT_ULL(6)
-#define NFS_ATTR_FATTR_PRESIZE		BIT_ULL(7)
-#define NFS_ATTR_FATTR_BLOCKS_USED	BIT_ULL(8)
-#define NFS_ATTR_FATTR_SPACE_USED	BIT_ULL(9)
-#define NFS_ATTR_FATTR_FSID		BIT_ULL(10)
-#define NFS_ATTR_FATTR_FILEID		BIT_ULL(11)
-#define NFS_ATTR_FATTR_ATIME		BIT_ULL(12)
-#define NFS_ATTR_FATTR_MTIME		BIT_ULL(13)
-#define NFS_ATTR_FATTR_CTIME		BIT_ULL(14)
-#define NFS_ATTR_FATTR_PREMTIME		BIT_ULL(15)
-#define NFS_ATTR_FATTR_PRECTIME		BIT_ULL(16)
-#define NFS_ATTR_FATTR_CHANGE		BIT_ULL(17)
-#define NFS_ATTR_FATTR_PRECHANGE	BIT_ULL(18)
-#define NFS_ATTR_FATTR_V4_LOCATIONS	BIT_ULL(19)
-#define NFS_ATTR_FATTR_V4_REFERRAL	BIT_ULL(20)
-#define NFS_ATTR_FATTR_MOUNTPOINT	BIT_ULL(21)
-#define NFS_ATTR_FATTR_MOUNTED_ON_FILEID BIT_ULL(22)
-#define NFS_ATTR_FATTR_OWNER_NAME	BIT_ULL(23)
-#define NFS_ATTR_FATTR_GROUP_NAME	BIT_ULL(24)
-#define NFS_ATTR_FATTR_V4_SECURITY_LABEL BIT_ULL(25)
-#define NFS_ATTR_FATTR_BTIME		BIT_ULL(26)
+#define NFS_ATTR_FATTR_TYPE		(1U << 0)
+#define NFS_ATTR_FATTR_MODE		(1U << 1)
+#define NFS_ATTR_FATTR_NLINK		(1U << 2)
+#define NFS_ATTR_FATTR_OWNER		(1U << 3)
+#define NFS_ATTR_FATTR_GROUP		(1U << 4)
+#define NFS_ATTR_FATTR_RDEV		(1U << 5)
+#define NFS_ATTR_FATTR_SIZE		(1U << 6)
+#define NFS_ATTR_FATTR_PRESIZE		(1U << 7)
+#define NFS_ATTR_FATTR_BLOCKS_USED	(1U << 8)
+#define NFS_ATTR_FATTR_SPACE_USED	(1U << 9)
+#define NFS_ATTR_FATTR_FSID		(1U << 10)
+#define NFS_ATTR_FATTR_FILEID		(1U << 11)
+#define NFS_ATTR_FATTR_ATIME		(1U << 12)
+#define NFS_ATTR_FATTR_MTIME		(1U << 13)
+#define NFS_ATTR_FATTR_CTIME		(1U << 14)
+#define NFS_ATTR_FATTR_PREMTIME		(1U << 15)
+#define NFS_ATTR_FATTR_PRECTIME		(1U << 16)
+#define NFS_ATTR_FATTR_CHANGE		(1U << 17)
+#define NFS_ATTR_FATTR_PRECHANGE	(1U << 18)
+#define NFS_ATTR_FATTR_V4_LOCATIONS	(1U << 19)
+#define NFS_ATTR_FATTR_V4_REFERRAL	(1U << 20)
+#define NFS_ATTR_FATTR_MOUNTPOINT	(1U << 21)
+#define NFS_ATTR_FATTR_MOUNTED_ON_FILEID (1U << 22)
+#define NFS_ATTR_FATTR_OWNER_NAME	(1U << 23)
+#define NFS_ATTR_FATTR_GROUP_NAME	(1U << 24)
+#define NFS_ATTR_FATTR_V4_SECURITY_LABEL (1U << 25)
 
 #define NFS_ATTR_FATTR (NFS_ATTR_FATTR_TYPE \
 		| NFS_ATTR_FATTR_MODE \
@@ -128,7 +126,6 @@
 		| NFS_ATTR_FATTR_SPACE_USED)
 #define NFS_ATTR_FATTR_V4 (NFS_ATTR_FATTR \
 		| NFS_ATTR_FATTR_SPACE_USED \
-		| NFS_ATTR_FATTR_BTIME \
 		| NFS_ATTR_FATTR_V4_SECURITY_LABEL)
 
 /*
@@ -209,7 +206,6 @@
 };
 
 struct nfs4_sequence_res {
-	const struct nfs4_sequence_slot_ops *sr_slot_ops;
 	struct nfs4_slot	*sr_slot;	/* slot used to send request */
 	unsigned long		sr_timestamp;
 	int			sr_status;	/* sequence operation status */
@@ -450,23 +446,7 @@
 
 struct stateowner_id {
 	__u64	create_time;
-	__u64	uniquifier;
-};
-
-struct nfs4_open_delegation {
-	__u32 open_delegation_type;
-	union {
-		struct {
-			fmode_t			type;
-			__u32			do_recall;
-			nfs4_stateid		stateid;
-			unsigned long		pagemod_limit;
-		};
-		struct {
-			__u32			why_no_delegation;
-			__u32			will_notify;
-		};
-	};
+	__u32	uniquifier;
 };
 
 /*
@@ -488,7 +468,7 @@
 			nfs4_verifier   verifier; /* EXCLUSIVE */
 		};
 		nfs4_stateid	delegation;		/* CLAIM_DELEGATE_CUR */
-		__u32		delegation_type;	/* CLAIM_PREVIOUS */
+		fmode_t		delegation_type;	/* CLAIM_PREVIOUS */
 	} u;
 	const struct qstr *	name;
 	const struct nfs_server *server;	 /* Needed for ID mapping */
@@ -510,10 +490,13 @@
 	struct nfs_fattr *      f_attr;
 	struct nfs_seqid *	seqid;
 	const struct nfs_server *server;
+	fmode_t			delegation_type;
+	nfs4_stateid		delegation;
+	unsigned long		pagemod_limit;
+	__u32			do_recall;
 	__u32			attrset[NFS4_BITMAP_SIZE];
 	struct nfs4_string	*owner;
 	struct nfs4_string	*group_owner;
-	struct nfs4_open_delegation	delegation;
 	__u32			access_request;
 	__u32			access_supported;
 	__u32			access_result;
@@ -626,13 +609,6 @@
 	struct nfs4_sequence_res	seq_res;
 };
 
-struct nfs4_delegattr {
-	struct timespec64	atime;
-	struct timespec64	mtime;
-	bool			atime_set;
-	bool			mtime_set;
-};
-
 struct nfs4_delegreturnargs {
 	struct nfs4_sequence_args	seq_args;
 	const struct nfs_fh *fhandle;
@@ -640,7 +616,6 @@
 	const u32 *bitmask;
 	u32 bitmask_store[NFS_BITMASK_SZ];
 	struct nfs4_layoutreturn_args *lr_args;
-	struct nfs4_delegattr *sattr_args;
 };
 
 struct nfs4_delegreturnres {
@@ -649,8 +624,6 @@
 	struct nfs_server *server;
 	struct nfs4_layoutreturn_res *lr_res;
 	int lr_ret;
-	bool sattr_res;
-	int sattr_ret;
 };
 
 /*
@@ -863,7 +836,7 @@
 	size_t				acl_len;
 	size_t				acl_data_offset;
 	int				acl_flags;
-	struct folio *			acl_scratch;
+	struct page *			acl_scratch;
 };
 
 struct nfs_setattrres {
@@ -1093,19 +1066,12 @@
 	struct nfs4_sequence_args	seq_args;
 	const struct nfs_fh *		fh;
 	const u32 *			bitmask;
-	bool				get_dir_deleg;
-};
-
-struct nfs4_gdd_res {
-	u32				status;
-	nfs4_stateid			deleg;
 };
 
 struct nfs4_getattr_res {
 	struct nfs4_sequence_res	seq_res;
 	const struct nfs_server *	server;
 	struct nfs_fattr *		fattr;
-	struct nfs4_gdd_res *		gdd_res;
 };
 
 struct nfs4_link_arg {
@@ -1224,14 +1190,6 @@
 	struct nfs_fsstat	       *fsstat;
 };
 
-struct nfs4_open_caps {
-	u32				oa_share_access[1];
-	u32				oa_share_deny[1];
-	u32				oa_share_access_want[1];
-	u32				oa_open_claim[1];
-	u32				oa_createmode[1];
-};
-
 struct nfs4_server_caps_arg {
 	struct nfs4_sequence_args	seq_args;
 	struct nfs_fh		       *fhandle;
@@ -1248,7 +1206,6 @@
 	u32				fh_expire_type;
 	u32				case_insensitive;
 	u32				case_preserving;
-	struct nfs4_open_caps		open_caps;
 };
 
 #define NFS4_PATHNAME_MAXCOMPONENTS 512
@@ -1324,6 +1281,15 @@
 	unsigned char			renew:1;
 };
 
+#endif /* CONFIG_NFS_V4 */
+
+struct nfstime4 {
+	u64	seconds;
+	u32	nseconds;
+};
+
+#ifdef CONFIG_NFS_V4_1
+
 struct pnfs_commit_bucket {
 	struct list_head written;
 	struct list_head committing;
@@ -1338,7 +1304,7 @@
 	struct rcu_head rcu;
 	refcount_t refcount;
 	unsigned int nbuckets;
-	struct pnfs_commit_bucket buckets[] __counted_by(nbuckets);
+	struct pnfs_commit_bucket buckets[];
 };
 
 struct pnfs_ds_commit_info {
@@ -1440,7 +1406,7 @@
 
 struct nfs41_test_stateid_args {
 	struct nfs4_sequence_args	seq_args;
-	nfs4_stateid			stateid;
+	nfs4_stateid			*stateid;
 };
 
 struct nfs41_test_stateid_res {
@@ -1463,7 +1429,7 @@
 struct pnfs_ds_commit_info {
 };
 
-#endif /* CONFIG_NFS_V4 */
+#endif /* CONFIG_NFS_V4_1 */
 
 #ifdef CONFIG_NFS_V4_2
 struct nfs42_falloc_args {
@@ -1522,9 +1488,8 @@
 
 struct nfs42_offload_status_res {
 	struct nfs4_sequence_res	osr_seq_res;
-	u64				osr_count;
-	int				complete_count;
-	u32				osr_complete;
+	uint64_t			osr_count;
+	int				osr_status;
 };
 
 struct nfs42_copy_notify_args {
@@ -1600,7 +1565,7 @@
 
 struct nfs42_listxattrsres {
 	struct nfs4_sequence_res	seq_res;
-	struct folio			*scratch;
+	struct page			*scratch;
 	void				*xattr_buf;
 	size_t				xattr_len;
 	u64				cookie;
@@ -1640,7 +1605,6 @@
 	NFS_IOHDR_RESEND_PNFS,
 	NFS_IOHDR_RESEND_MDS,
 	NFS_IOHDR_UNSTABLE_WRITES,
-	NFS_IOHDR_ODIRECT,
 };
 
 struct nfs_io_completion;
@@ -1663,7 +1627,6 @@
 	void			*netfs;
 #endif
 
-	unsigned short		retrans;
 	int			pnfs_error;
 	int			error;		/* merge with pnfs_error */
 	unsigned int		good_bytes;	/* boundary of good data */
@@ -1762,6 +1725,14 @@
 	bool cancelled;
 };
 
+#if IS_ENABLED(CONFIG_ENFS)
+struct nfs_extend_xdr_arg {
+	int	 maxsize;
+	int	 buflen;
+	char	*pBuf;
+};
+#endif
+
 struct nfs_access_entry;
 struct nfs_client;
 struct rpc_timeout;
@@ -1790,7 +1761,7 @@
 			    struct nfs_fattr *, struct inode *);
 	int	(*setattr) (struct dentry *, struct nfs_fattr *,
 			    struct iattr *);
-	int	(*lookup)  (struct inode *, struct dentry *, const struct qstr *,
+	int	(*lookup)  (struct inode *, struct dentry *,
 			    struct nfs_fh *, struct nfs_fattr *);
 	int	(*lookupp) (struct inode *, struct nfs_fh *,
 			    struct nfs_fattr *);
@@ -1805,14 +1776,13 @@
 	int	(*unlink_done) (struct rpc_task *, struct inode *);
 	void	(*rename_setup)  (struct rpc_message *msg,
 			struct dentry *old_dentry,
-			struct dentry *new_dentry,
-			struct inode *same_parent);
+			struct dentry *new_dentry);
 	void	(*rename_rpc_prepare)(struct rpc_task *task, struct nfs_renamedata *);
 	int	(*rename_done) (struct rpc_task *task, struct inode *old_dir, struct inode *new_dir);
 	int	(*link)    (struct inode *, struct inode *, const struct qstr *);
-	int	(*symlink) (struct inode *, struct dentry *, struct folio *,
+	int	(*symlink) (struct inode *, struct dentry *, struct page *,
 			    unsigned int, struct iattr *);
-	struct dentry *(*mkdir)   (struct inode *, struct dentry *, struct iattr *);
+	int	(*mkdir)   (struct inode *, struct dentry *, struct iattr *);
 	int	(*rmdir)   (struct inode *, const struct qstr *);
 	int	(*readdir) (struct nfs_readdir_arg *, struct nfs_readdir_res *);
 	int	(*mknod)   (struct inode *, struct dentry *, struct iattr *,
@@ -1845,8 +1815,7 @@
 				int open_flags,
 				struct iattr *iattr,
 				int *);
-	int (*have_delegation)(struct inode *, fmode_t, int);
-	void (*return_delegation)(struct inode *);
+	int (*have_delegation)(struct inode *, fmode_t);
 	struct nfs_client *(*alloc_client) (const struct nfs_client_initdata *);
 	struct nfs_client *(*init_client) (struct nfs_client *,
 				const struct nfs_client_initdata *);
@@ -1860,22 +1829,11 @@
 };
 
 /*
- * Helper functions used by NFS client and/or server
+ * 	NFS_CALL(getattr, inode, (fattr));
+ * into
+ *	NFS_PROTO(inode)->getattr(fattr);
  */
-static inline void encode_opaque_fixed(struct xdr_stream *xdr,
-				       const void *buf, size_t len)
-{
-	WARN_ON_ONCE(xdr_stream_encode_opaque_fixed(xdr, buf, len) < 0);
-}
-
-static inline int decode_opaque_fixed(struct xdr_stream *xdr,
-				      void *buf, size_t len)
-{
-	ssize_t ret = xdr_stream_decode_opaque_fixed(xdr, buf, len);
-	if (unlikely(ret < 0))
-		return -EIO;
-	return 0;
-}
+#define NFS_CALL(op, inode, args)	NFS_PROTO(inode)->op args
 
 /*
  * Function vectors etc. for the NFS client
@@ -1890,4 +1848,4 @@
 extern const struct rpc_version nfsacl_version3;
 extern const struct rpc_program nfsacl_program;
 
-#endif /* _LINUX_NFS_XDR_H */
+#endif
