--- /tmp/tmp.tYSqORsSZN/ubuntu	2026-04-30 18:37:01.570059288 +0200
+++ /home/darren/development/enfs/vendor/openeuler/include/linux/sunrpc/sched.h	2026-04-30 17:42:52.000000000 +0200
@@ -17,6 +17,7 @@
 #include <linux/wait_bit.h>
 #include <linux/workqueue.h>
 #include <linux/sunrpc/xdr.h>
+#include <linux/kabi.h>
 
 /*
  * This is the actual RPC procedure call info.
@@ -27,6 +28,9 @@
 	void *			rpc_argp;	/* Arguments */
 	void *			rpc_resp;	/* Result */
 	const struct cred *	rpc_cred;	/* Credentials */
+
+	KABI_RESERVE(1)
+	KABI_RESERVE(2)
 };
 
 struct rpc_call_ops;
@@ -38,17 +42,6 @@
 };
 
 /*
- * This describes a timeout strategy
- */
-struct rpc_timeout {
-	unsigned long		to_initval,		/* initial timeout */
-				to_maxval,		/* max timeout */
-				to_increment;		/* if !exponential */
-	unsigned int		to_retries;		/* max # of retries */
-	unsigned char		to_exponential;
-};
-
-/*
  * This is the RPC task struct
  */
 struct rpc_task {
@@ -102,6 +95,11 @@
 	unsigned char		tk_priority : 2,/* Task priority */
 				tk_garb_retry : 2,
 				tk_cred_retry : 2;
+
+	KABI_RESERVE(1)
+	KABI_RESERVE(2)
+	KABI_RESERVE(3)
+	KABI_RESERVE(4)
 };
 
 typedef void			(*rpc_action)(struct rpc_task *);
@@ -111,6 +109,9 @@
 	void (*rpc_call_done)(struct rpc_task *, void *);
 	void (*rpc_count_stats)(struct rpc_task *, void *);
 	void (*rpc_release)(void *);
+
+	KABI_RESERVE(1)
+	KABI_RESERVE(2)
 };
 
 struct rpc_task_setup {
@@ -124,6 +125,9 @@
 	struct workqueue_struct *workqueue;
 	unsigned short flags;
 	signed char priority;
+
+	KABI_RESERVE(1)
+	KABI_RESERVE(2)
 };
 
 /*
@@ -134,7 +138,6 @@
 #define RPC_TASK_MOVEABLE	0x0004		/* nfs4.1+ rpc tasks */
 #define RPC_TASK_NULLCREDS	0x0010		/* Use AUTH_NULL credential */
 #define RPC_CALL_MAJORSEEN	0x0020		/* major timeout seen */
-#define RPC_TASK_NETUNREACH_FATAL 0x0040	/* ENETUNREACH is fatal */
 #define RPC_TASK_DYNAMIC	0x0080		/* task was kmalloc'ed */
 #define	RPC_TASK_NO_ROUND_ROBIN	0x0100		/* send requests on "main" xprt */
 #define RPC_TASK_SOFT		0x0200		/* Use soft timeouts */
@@ -144,7 +147,10 @@
 #define RPC_TASK_NOCONNECT	0x2000		/* return ENOTCONN if not connected */
 #define RPC_TASK_NO_RETRANS_TIMEOUT	0x4000		/* wait forever for a reply */
 #define RPC_TASK_CRED_NOREF	0x8000		/* No refcount on the credential */
-
+#if IS_ENABLED(CONFIG_SUNRPC_ENFS)
+#define RPC_TASK_FIXED		0x0040
+#define RPC_TASK_ENFS		0x0008	  /* enfs rpc program task */
+#endif
 #define RPC_IS_ASYNC(t)		((t)->tk_flags & RPC_TASK_ASYNC)
 #define RPC_IS_SWAPPER(t)	((t)->tk_flags & RPC_TASK_SWAPPER)
 #define RPC_IS_SOFT(t)		((t)->tk_flags & (RPC_TASK_SOFT|RPC_TASK_TIMEOUT))
@@ -218,8 +224,7 @@
  */
 struct rpc_task *rpc_new_task(const struct rpc_task_setup *);
 struct rpc_task *rpc_run_task(const struct rpc_task_setup *);
-struct rpc_task *rpc_run_bc_task(struct rpc_rqst *req,
-		struct rpc_timeout *timeout);
+struct rpc_task *rpc_run_bc_task(struct rpc_rqst *req);
 void		rpc_put_task(struct rpc_task *);
 void		rpc_put_task_async(struct rpc_task *);
 bool		rpc_task_set_rpc_status(struct rpc_task *task, int rpc_status);
