--- /tmp/tmp.tYSqORsSZN/ubuntu	2026-04-30 18:37:05.926058534 +0200
+++ /home/darren/development/enfs/vendor/openeuler/net/sunrpc/Kconfig	2026-04-30 17:42:53.000000000 +0200
@@ -101,20 +101,6 @@
 
 	  If unsure, say Y.
 
-config SUNRPC_DEBUG_TRACE
-	bool "RPC: Send dfprintk() output to the trace buffer"
-	depends on SUNRPC_DEBUG && TRACING
-	default n
-	help
-          dprintk() output can be voluminous, which can overwhelm the
-          kernel's logging facility as it must be sent to the console.
-          This option causes dprintk() output to go to the trace buffer
-          instead of the kernel log.
-
-          This will cause warnings about trace_printk() being used to be
-          logged at boot time, so say N unless you are debugging a problem
-          with sunrpc-based clients or services.
-
 config SUNRPC_XPRT_RDMA
 	tristate "RPC-over-RDMA transport"
 	depends on SUNRPC && INFINIBAND && INFINIBAND_ADDR_TRANS
@@ -129,3 +115,15 @@
 
 	  If unsure, or you know there is no RDMA capability on your
 	  hardware platform, say N.
+
+config SUNRPC_ENFS
+	bool "sunrpc support ENFS"
+	depends on SUNRPC
+	depends on X86 || X86_64 || ARM64
+	default n
+	help
+	 This option enables support multipath of the NFS protocol
+	 in the kernel's NFS client.
+	 This feature will improve performance and reliability.
+
+	 If sure, say Y.
