--- /tmp/tmp.tYSqORsSZN/ubuntu	2026-04-30 18:37:14.850056887 +0200
+++ /home/darren/development/enfs/vendor/openeuler/net/sunrpc/xprt.c	2026-04-30 17:42:53.000000000 +0200
@@ -51,6 +51,7 @@
 #include <linux/sunrpc/bc_xprt.h>
 #include <linux/rcupdate.h>
 #include <linux/sched/mm.h>
+#include <linux/sunrpc/sunrpc_enfs_adapter.h>
 
 #include <trace/events/sunrpc.h>
 
@@ -282,8 +283,9 @@
 out_unlock:
 	xprt_clear_locked(xprt);
 out_sleep:
+	rpc_multipath_ops_adjust_task_timeout(task, NULL);
 	task->tk_status = -EAGAIN;
-	if (RPC_IS_SOFT(task) || RPC_IS_SOFTCONN(task))
+	if  (RPC_IS_SOFT(task))
 		rpc_sleep_on_timeout(&xprt->sending, task, NULL,
 				xprt_request_timeout(req));
 	else
@@ -348,8 +350,10 @@
 out_unlock:
 	xprt_clear_locked(xprt);
 out_sleep:
+	rpc_multipath_ops_adjust_task_timeout(task, NULL);
+
 	task->tk_status = -EAGAIN;
-	if (RPC_IS_SOFT(task) || RPC_IS_SOFTCONN(task))
+	if (RPC_IS_SOFT(task))
 		rpc_sleep_on_timeout(&xprt->sending, task, NULL,
 				xprt_request_timeout(req));
 	else
@@ -608,6 +612,7 @@
  */
 void xprt_wait_for_buffer_space(struct rpc_xprt *xprt)
 {
+	rpc_multipath_ops_adjust_task_timeout(xprt->snd_task, NULL);
 	set_bit(XPRT_WRITE_SPACE, &xprt->state);
 }
 EXPORT_SYMBOL_GPL(xprt_wait_for_buffer_space);
@@ -651,9 +656,9 @@
 		jiffies + nsecs_to_jiffies(-delta);
 }
 
-static unsigned long xprt_calc_majortimeo(struct rpc_rqst *req,
-		const struct rpc_timeout *to)
+static unsigned long xprt_calc_majortimeo(struct rpc_rqst *req)
 {
+	const struct rpc_timeout *to = req->rq_task->tk_client->cl_timeout;
 	unsigned long majortimeo = req->rq_timeout;
 
 	if (to->to_exponential)
@@ -665,10 +670,9 @@
 	return majortimeo;
 }
 
-static void xprt_reset_majortimeo(struct rpc_rqst *req,
-		const struct rpc_timeout *to)
+static void xprt_reset_majortimeo(struct rpc_rqst *req)
 {
-	req->rq_majortimeo += xprt_calc_majortimeo(req, to);
+	req->rq_majortimeo += xprt_calc_majortimeo(req);
 }
 
 static void xprt_reset_minortimeo(struct rpc_rqst *req)
@@ -676,8 +680,7 @@
 	req->rq_minortimeo += req->rq_timeout;
 }
 
-static void xprt_init_majortimeo(struct rpc_task *task, struct rpc_rqst *req,
-		const struct rpc_timeout *to)
+static void xprt_init_majortimeo(struct rpc_task *task, struct rpc_rqst *req)
 {
 	unsigned long time_init;
 	struct rpc_xprt *xprt = req->rq_xprt;
@@ -686,9 +689,8 @@
 		time_init = jiffies;
 	else
 		time_init = xprt_abs_ktime_to_jiffies(task->tk_start);
-
-	req->rq_timeout = to->to_initval;
-	req->rq_majortimeo = time_init + xprt_calc_majortimeo(req, to);
+	req->rq_timeout = task->tk_client->cl_timeout->to_initval;
+	req->rq_majortimeo = time_init + xprt_calc_majortimeo(req);
 	req->rq_minortimeo = time_init + req->rq_timeout;
 }
 
@@ -716,7 +718,7 @@
 	} else {
 		req->rq_timeout = to->to_initval;
 		req->rq_retries = 0;
-		xprt_reset_majortimeo(req, to);
+		xprt_reset_majortimeo(req);
 		/* Reset the RTT counters == "slow start" */
 		spin_lock(&xprt->transport_lock);
 		rpc_init_rtt(req->rq_task->tk_client->cl_rtt, to->to_initval);
@@ -854,7 +856,7 @@
 static void
 xprt_init_autodisconnect(struct timer_list *t)
 {
-	struct rpc_xprt *xprt = timer_container_of(xprt, t, timer);
+	struct rpc_xprt *xprt = from_timer(xprt, t, timer);
 
 	if (!RB_EMPTY_ROOT(&xprt->recv_queue))
 		return;
@@ -1167,7 +1169,7 @@
 	spin_unlock(&xprt->queue_lock);
 
 	/* Turn off autodisconnect */
-	timer_delete_sync(&xprt->timer);
+	del_timer_sync(&xprt->timer);
 	return 0;
 }
 
@@ -1365,7 +1367,7 @@
 				INIT_LIST_HEAD(&req->rq_xmit2);
 				goto out;
 			}
-		} else if (req->rq_seqno_count == 0) {
+		} else if (!req->rq_seqno) {
 			list_for_each_entry(pos, &xprt->xmit_queue, rq_xmit) {
 				if (pos->rq_task->tk_owner != task->tk_owner)
 					continue;
@@ -1398,12 +1400,6 @@
 	if (!test_and_clear_bit(RPC_TASK_NEED_XMIT, &task->tk_runstate))
 		return;
 	if (!list_empty(&req->rq_xmit)) {
-		struct rpc_xprt *xprt = req->rq_xprt;
-
-		if (list_is_first(&req->rq_xmit, &xprt->xmit_queue) &&
-		    xprt->ops->abort_send_request)
-			xprt->ops->abort_send_request(req);
-
 		list_del(&req->rq_xmit);
 		if (!list_empty(&req->rq_xmit2)) {
 			struct rpc_rqst *next = list_first_entry(&req->rq_xmit2,
@@ -1547,9 +1543,6 @@
 	int is_retrans = RPC_WAS_SENT(task);
 	int status;
 
-	if (test_bit(XPRT_CLOSE_WAIT, &xprt->state))
-		return -ENOTCONN;
-
 	if (!req->rq_bytes_sent) {
 		if (xprt_request_data_received(task)) {
 			status = 0;
@@ -1709,7 +1702,7 @@
 		goto out;
 	++xprt->num_reqs;
 	spin_unlock(&xprt->reserve_lock);
-	req = kzalloc_obj(*req, rpc_task_gfp_mask());
+	req = kzalloc(sizeof(*req), rpc_task_gfp_mask());
 	spin_lock(&xprt->reserve_lock);
 	if (req != NULL)
 		goto out;
@@ -1829,7 +1822,7 @@
 	xprt_init(xprt, net);
 
 	for (i = 0; i < num_prealloc; i++) {
-		req = kzalloc_obj(struct rpc_rqst);
+		req = kzalloc(sizeof(struct rpc_rqst), GFP_KERNEL);
 		if (!req)
 			goto out_free;
 		list_add(&req->rq_list, &xprt->free);
@@ -1898,8 +1891,9 @@
 	req->rq_snd_buf.bvec = NULL;
 	req->rq_rcv_buf.bvec = NULL;
 	req->rq_release_snd_buf = NULL;
-	req->rq_seqno_count = 0;
-	xprt_init_majortimeo(task, req, task->tk_client->cl_timeout);
+	xprt_init_majortimeo(task, req);
+
+	rpc_multipath_ops_init_task_req(task, req);
 
 	trace_xprt_reserve(req);
 }
@@ -1974,6 +1968,9 @@
 
 	xprt = req->rq_xprt;
 	xprt_request_dequeue_xprt(task);
+
+	rpc_multipath_ops_xprt_iostat(task);
+
 	spin_lock(&xprt->transport_lock);
 	xprt->ops->release_xprt(xprt, task);
 	if (xprt->ops->release_request)
@@ -1993,11 +1990,11 @@
 	else
 		xprt_free_bc_request(req);
 }
+EXPORT_SYMBOL_GPL(xprt_release);
 
 #ifdef CONFIG_SUNRPC_BACKCHANNEL
 void
-xprt_init_bc_request(struct rpc_rqst *req, struct rpc_task *task,
-		const struct rpc_timeout *to)
+xprt_init_bc_request(struct rpc_rqst *req, struct rpc_task *task)
 {
 	struct xdr_buf *xbufp = &req->rq_snd_buf;
 
@@ -2010,13 +2007,6 @@
 	 */
 	xbufp->len = xbufp->head[0].iov_len + xbufp->page_len +
 		xbufp->tail[0].iov_len;
-	/*
-	 * Backchannel Replies are sent with !RPC_TASK_SOFT and
-	 * RPC_TASK_NO_RETRANS_TIMEOUT. The major timeout setting
-	 * affects only how long each Reply waits to be sent when
-	 * a transport connection cannot be established.
-	 */
-	xprt_init_majortimeo(task, req, to);
 }
 #endif
 
@@ -2084,12 +2074,23 @@
 		xprt_destroy(xprt);
 		return ERR_PTR(-EINVAL);
 	}
+#if IS_ENABLED(CONFIG_SUNRPC_ENFS)
+	xprt->servername = rpc_multipath_set_servername(args->servername, GFP_KERNEL);
+#else
 	xprt->servername = kstrdup(args->servername, GFP_KERNEL);
+#endif
 	if (xprt->servername == NULL) {
 		xprt_destroy(xprt);
 		return ERR_PTR(-ENOMEM);
 	}
 
+#if IS_ENABLED(CONFIG_SUNRPC_ENFS)
+	if (!rpc_multipath_ops_create_xprt(xprt)) {
+		xprt_destroy(xprt);
+		return ERR_PTR(-ENOMEM);
+	}
+#endif
+
 	rpc_xprt_debugfs_register(xprt);
 
 	trace_xprt_create(xprt);
@@ -2109,7 +2110,11 @@
 	rpc_destroy_wait_queue(&xprt->pending);
 	rpc_destroy_wait_queue(&xprt->sending);
 	rpc_destroy_wait_queue(&xprt->backlog);
+#if IS_ENABLED(CONFIG_SUNRPC_ENFS)
+	rpc_multipath_free_servername(xprt);
+#else
 	kfree(xprt->servername);
+#endif
 	/*
 	 * Destroy any existing back channel
 	 */
@@ -2139,7 +2144,7 @@
 	 * can only run *before* del_time_sync(), never after.
 	 */
 	spin_lock(&xprt->transport_lock);
-	timer_delete_sync(&xprt->timer);
+	del_timer_sync(&xprt->timer);
 	spin_unlock(&xprt->transport_lock);
 
 	/*
