--- /tmp/tmp.tYSqORsSZN/ubuntu	2026-04-30 18:37:10.418057727 +0200
+++ /home/darren/development/enfs/vendor/openeuler/net/sunrpc/clnt.c	2026-04-30 17:42:53.000000000 +0200
@@ -38,6 +38,7 @@
 #include <linux/sunrpc/rpc_pipe_fs.h>
 #include <linux/sunrpc/metrics.h>
 #include <linux/sunrpc/bc_xprt.h>
+#include <linux/sunrpc/sunrpc_enfs_adapter.h>
 #include <trace/events/sunrpc.h>
 
 #include "sunrpc.h"
@@ -48,8 +49,13 @@
 # define RPCDBG_FACILITY	RPCDBG_CALL
 #endif
 
+/*
+ * All RPC clients are linked into this list
+ */
+
 static DECLARE_WAIT_QUEUE_HEAD(destroy_wait);
 
+
 static void	call_start(struct rpc_task *task);
 static void	call_reserve(struct rpc_task *task);
 static void	call_reserveresult(struct rpc_task *task);
@@ -72,7 +78,7 @@
 				  struct xdr_stream *xdr);
 static int	rpc_ping(struct rpc_clnt *clnt);
 static int	rpc_ping_noreply(struct rpc_clnt *clnt);
-static void	rpc_check_timeout(struct rpc_task *task);
+static void	rpc_check_timeout(struct rpc_task *task, bool failover);
 
 static void rpc_register_client(struct rpc_clnt *clnt)
 {
@@ -112,46 +118,47 @@
 	}
 }
 
-static int rpc_setup_pipedir_sb(struct super_block *sb,
+static struct dentry *rpc_setup_pipedir_sb(struct super_block *sb,
 				    struct rpc_clnt *clnt)
 {
 	static uint32_t clntid;
 	const char *dir_name = clnt->cl_program->pipe_dir_name;
 	char name[15];
-	struct dentry *dir;
-	int err;
+	struct dentry *dir, *dentry;
 
 	dir = rpc_d_lookup_sb(sb, dir_name);
 	if (dir == NULL) {
 		pr_info("RPC: pipefs directory doesn't exist: %s\n", dir_name);
-		return -ENOENT;
+		return dir;
 	}
 	for (;;) {
 		snprintf(name, sizeof(name), "clnt%x", (unsigned int)clntid++);
 		name[sizeof(name) - 1] = '\0';
-		err = rpc_create_client_dir(dir, name, clnt);
-		if (!err)
+		dentry = rpc_create_client_dir(dir, name, clnt);
+		if (!IS_ERR(dentry))
 			break;
-		if (err == -EEXIST)
+		if (dentry == ERR_PTR(-EEXIST))
 			continue;
 		printk(KERN_INFO "RPC: Couldn't create pipefs entry"
-				" %s/%s, error %d\n",
-				dir_name, name, err);
+				" %s/%s, error %ld\n",
+				dir_name, name, PTR_ERR(dentry));
 		break;
 	}
 	dput(dir);
-	return err;
+	return dentry;
 }
 
 static int
 rpc_setup_pipedir(struct super_block *pipefs_sb, struct rpc_clnt *clnt)
 {
+	struct dentry *dentry;
+
 	clnt->pipefs_sb = pipefs_sb;
 
 	if (clnt->cl_program->pipe_dir_name != NULL) {
-		int err = rpc_setup_pipedir_sb(pipefs_sb, clnt);
-		if (err && err != -ENOENT)
-			return err;
+		dentry = rpc_setup_pipedir_sb(pipefs_sb, clnt);
+		if (IS_ERR(dentry))
+			return PTR_ERR(dentry);
 	}
 	return 0;
 }
@@ -179,9 +186,16 @@
 static int __rpc_clnt_handle_event(struct rpc_clnt *clnt, unsigned long event,
 				   struct super_block *sb)
 {
+	struct dentry *dentry;
+
 	switch (event) {
 	case RPC_PIPEFS_MOUNT:
-		return rpc_setup_pipedir_sb(sb, clnt);
+		dentry = rpc_setup_pipedir_sb(sb, clnt);
+		if (!dentry)
+			return -ENOENT;
+		if (IS_ERR(dentry))
+			return PTR_ERR(dentry);
+		break;
 	case RPC_PIPEFS_UMOUNT:
 		__rpc_clnt_remove_pipedir(clnt);
 		break;
@@ -262,6 +276,9 @@
 	old = rcu_dereference_protected(clnt->cl_xprt,
 			lockdep_is_held(&clnt->cl_lock));
 
+	if (!xprt_bound(xprt))
+		clnt->cl_autobind = 1;
+
 	clnt->cl_timeout = timeout;
 	rcu_assign_pointer(clnt->cl_xprt, xprt);
 	spin_unlock(&clnt->cl_lock);
@@ -271,14 +288,8 @@
 
 static void rpc_clnt_set_nodename(struct rpc_clnt *clnt, const char *nodename)
 {
-	ssize_t copied;
-
-	copied = strscpy(clnt->cl_nodename,
-			 nodename, sizeof(clnt->cl_nodename));
-
-	clnt->cl_nodelen = copied < 0
-				? sizeof(clnt->cl_nodename) - 1
-				: copied;
+	clnt->cl_nodelen = strlcpy(clnt->cl_nodename,
+			nodename, sizeof(clnt->cl_nodename));
 }
 
 static int rpc_client_register(struct rpc_clnt *clnt,
@@ -374,7 +385,7 @@
 		goto out_err;
 
 	err = -ENOMEM;
-	clnt = kzalloc_obj(*clnt);
+	clnt = kzalloc(sizeof(*clnt), GFP_KERNEL);
 	if (!clnt)
 		goto out_err;
 	clnt->cl_parent = parent ? : clnt;
@@ -486,6 +497,8 @@
 		}
 	}
 
+	rpc_multipath_ops_create_clnt(args, clnt);
+
 	clnt->cl_softrtry = 1;
 	if (args->flags & (RPC_CLNT_CREATE_HARDRTRY|RPC_CLNT_CREATE_SOFTERR)) {
 		clnt->cl_softrtry = 0;
@@ -501,8 +514,6 @@
 		clnt->cl_discrtry = 1;
 	if (!(args->flags & RPC_CLNT_CREATE_QUIET))
 		clnt->cl_chatty = 1;
-	if (args->flags & RPC_CLNT_CREATE_NETUNREACH_FATAL)
-		clnt->cl_netunreach_fatal = 1;
 
 	return clnt;
 }
@@ -532,7 +543,7 @@
 		.connect_timeout = args->connect_timeout,
 		.reconnect_timeout = args->reconnect_timeout,
 	};
-	char servername[RPC_MAXNETNAMELEN];
+	char servername[48];
 	struct rpc_clnt *clnt;
 	int i;
 
@@ -653,7 +664,6 @@
 	new->cl_noretranstimeo = clnt->cl_noretranstimeo;
 	new->cl_discrtry = clnt->cl_discrtry;
 	new->cl_chatty = clnt->cl_chatty;
-	new->cl_netunreach_fatal = clnt->cl_netunreach_fatal;
 	new->cl_principal = clnt->cl_principal;
 	new->cl_max_connect = clnt->cl_max_connect;
 	return new;
@@ -792,24 +802,15 @@
 }
 EXPORT_SYMBOL_GPL(rpc_switch_client_transport);
 
-static struct rpc_xprt_switch *rpc_clnt_xprt_switch_get(struct rpc_clnt *clnt)
-{
-	struct rpc_xprt_switch *xps;
-
-	rcu_read_lock();
-	xps = xprt_switch_get(rcu_dereference(clnt->cl_xpi.xpi_xpswitch));
-	rcu_read_unlock();
-
-	return xps;
-}
-
 static
 int _rpc_clnt_xprt_iter_init(struct rpc_clnt *clnt, struct rpc_xprt_iter *xpi,
 			     void func(struct rpc_xprt_iter *xpi, struct rpc_xprt_switch *xps))
 {
 	struct rpc_xprt_switch *xps;
 
-	xps = rpc_clnt_xprt_switch_get(clnt);
+	rcu_read_lock();
+	xps = xprt_switch_get(rcu_dereference(clnt->cl_xpi.xpi_xpswitch));
+	rcu_read_unlock();
 	if (xps == NULL)
 		return -EAGAIN;
 	func(xpi, xps);
@@ -950,16 +951,13 @@
 
 	trace_rpc_clnt_shutdown(clnt);
 
-	clnt->cl_shutdown = 1;
 	while (!list_empty(&clnt->cl_tasks)) {
 		rpc_killall_tasks(clnt);
 		wait_event_timeout(destroy_wait,
 			list_empty(&clnt->cl_tasks), 1*HZ);
 	}
 
-	/* wait for tasks still in workqueue or waitqueue */
-	wait_event_timeout(destroy_wait,
-			   atomic_read(&clnt->cl_task_count) == 0, 1 * HZ);
+	rpc_multipath_ops_releas_clnt(clnt);
 
 	rpc_release_client(clnt);
 }
@@ -1088,6 +1086,8 @@
 
 	if (!xprt)
 		return NULL;
+
+	rpc_multipath_ops_inc_queuelen(xprt);
 	rcu_read_lock();
 	xps = rcu_dereference(clnt->cl_xpi.xpi_xpswitch);
 	atomic_long_inc(&xps->xps_queuelen);
@@ -1102,6 +1102,7 @@
 {
 	struct rpc_xprt_switch *xps;
 
+	rpc_multipath_ops_dec_queuelen(xprt);
 	atomic_long_dec(&xprt->queuelen);
 	rcu_read_lock();
 	xps = rcu_dereference(clnt->cl_xpi.xpi_xpswitch);
@@ -1136,7 +1137,6 @@
 		list_del(&task->tk_task);
 		spin_unlock(&clnt->cl_lock);
 		task->tk_client = NULL;
-		atomic_dec(&clnt->cl_task_count);
 
 		rpc_release_client(clnt);
 	}
@@ -1153,15 +1153,19 @@
 	return rpc_task_get_xprt(clnt, xprt);
 }
 
-static struct rpc_xprt *
+struct rpc_xprt *
 rpc_task_get_next_xprt(struct rpc_clnt *clnt)
 {
 	return rpc_task_get_xprt(clnt, xprt_iter_get_next(&clnt->cl_xpi));
 }
+EXPORT_SYMBOL_GPL(rpc_task_get_next_xprt);
 
 static
 void rpc_task_set_transport(struct rpc_task *task, struct rpc_clnt *clnt)
 {
+	if (task->tk_msg.rpc_proc)
+		rpc_multipath_ops_set_transport(task, clnt);
+
 	if (task->tk_xprt) {
 		if (!(test_bit(XPRT_OFFLINE, &task->tk_xprt->state) &&
 		      (task->tk_flags & RPC_TASK_MOVEABLE)))
@@ -1187,9 +1191,10 @@
 		task->tk_flags |= RPC_TASK_TIMEOUT;
 	if (clnt->cl_noretranstimeo)
 		task->tk_flags |= RPC_TASK_NO_RETRANS_TIMEOUT;
-	if (clnt->cl_netunreach_fatal)
-		task->tk_flags |= RPC_TASK_NETUNREACH_FATAL;
-	atomic_inc(&clnt->cl_task_count);
+	/* Add to the client's list of all tasks */
+	spin_lock(&clnt->cl_lock);
+	list_add_tail(&task->tk_task, &clnt->cl_tasks);
+	spin_unlock(&clnt->cl_lock);
 }
 
 static void
@@ -1313,10 +1318,8 @@
  * rpc_run_bc_task - Allocate a new RPC task for backchannel use, then run
  * rpc_execute against it
  * @req: RPC request
- * @timeout: timeout values to use for this task
  */
-struct rpc_task *rpc_run_bc_task(struct rpc_rqst *req,
-		struct rpc_timeout *timeout)
+struct rpc_task *rpc_run_bc_task(struct rpc_rqst *req)
 {
 	struct rpc_task *task;
 	struct rpc_task_setup task_setup_data = {
@@ -1335,7 +1338,7 @@
 		return task;
 	}
 
-	xprt_init_bc_request(req, task, timeout);
+	xprt_init_bc_request(req, task);
 
 	task->tk_action = call_bc_encode;
 	atomic_inc(&task->tk_count);
@@ -1457,12 +1460,12 @@
 	switch (sap->sa_family) {
 	case AF_INET:
 		err = kernel_bind(sock,
-				(struct sockaddr_unsized *)&rpc_inaddr_loopback,
+				(struct sockaddr *)&rpc_inaddr_loopback,
 				sizeof(rpc_inaddr_loopback));
 		break;
 	case AF_INET6:
 		err = kernel_bind(sock,
-				(struct sockaddr_unsized *)&rpc_in6addr_loopback,
+				(struct sockaddr *)&rpc_in6addr_loopback,
 				sizeof(rpc_in6addr_loopback));
 		break;
 	default:
@@ -1474,7 +1477,7 @@
 		goto out_release;
 	}
 
-	err = kernel_connect(sock, (struct sockaddr_unsized *)sap, salen, 0);
+	err = kernel_connect(sock, sap, salen, 0);
 	if (err < 0) {
 		dprintk("RPC:       can't connect UDP socket (%d)\n", err);
 		goto out_release;
@@ -1573,6 +1576,33 @@
 }
 EXPORT_SYMBOL_GPL(rpc_localaddr);
 
+int rpc_localalladdr(struct rpc_xprt *xprt, struct sockaddr *buf, size_t buflen)
+{
+	struct sockaddr_storage address;
+	struct sockaddr *sap = (struct sockaddr *)&address;
+	struct rpc_xprt *xpr;
+	struct net *net;
+	size_t salen;
+	int err;
+
+	rcu_read_lock();
+	xpr = rcu_dereference(xprt);
+	salen = xpr->addrlen;
+	memcpy(sap, &xpr->addr, salen);
+	net = get_net(xpr->xprt_net);
+	dprintk("NFS:net:%p\n", xpr->xprt_net);
+	rcu_read_unlock();
+
+	rpc_set_port(sap, 0);
+	err = rpc_sockname(net, sap, salen, buf);
+	put_net(net);
+	if (err != 0)
+		/* Couldn't discover local address, return ANYADDR */
+		return rpc_anyaddr(sap->sa_family, buf, buflen);
+	return 0;
+}
+EXPORT_SYMBOL_GPL(rpc_localalladdr);
+
 void
 rpc_setbufsize(struct rpc_clnt *clnt, unsigned int sndsize, unsigned int rcvsize)
 {
@@ -1784,14 +1814,9 @@
 	if (status >= 0) {
 		if (task->tk_rqstp) {
 			task->tk_action = call_refresh;
-
-			/* Add to the client's list of all tasks */
-			spin_lock(&task->tk_client->cl_lock);
-			if (list_empty(&task->tk_task))
-				list_add_tail(&task->tk_task, &task->tk_client->cl_tasks);
-			spin_unlock(&task->tk_client->cl_lock);
 			return;
 		}
+
 		rpc_call_rpcerror(task, -EIO);
 		return;
 	}
@@ -1803,6 +1828,14 @@
 	case -EAGAIN:	/* woken up; retry */
 		task->tk_action = call_retry_reserve;
 		return;
+#if IS_ENABLED(CONFIG_SUNRPC_ENFS)
+	case -ETIMEDOUT:	/* woken up; restart */
+		if (rpc_multipath_ops_task_need_call_start_again(task)) {
+			rpc_task_release_transport(task);
+			task->tk_action = call_start;
+		}
+		return;
+#endif
 	default:
 		rpc_call_rpcerror(task, status);
 	}
@@ -1856,13 +1889,13 @@
 		fallthrough;
 	case -EAGAIN:
 		status = -EACCES;
+		fallthrough;
+	case -EKEYEXPIRED:
 		if (!task->tk_cred_retry)
 			break;
 		task->tk_cred_retry--;
 		trace_rpc_retry_refresh_status(task);
 		return;
-	case -EKEYEXPIRED:
-		break;
 	case -ENOMEM:
 		rpc_delay(task, HZ >> 4);
 		return;
@@ -1890,6 +1923,12 @@
 	if (req->rq_buffer)
 		return;
 
+	if (proc->p_proc != 0) {
+		BUG_ON(proc->p_arglen == 0);
+		if (proc->p_decode != NULL)
+			BUG_ON(proc->p_replen == 0);
+	}
+
 	/*
 	 * Calculate the size (in quads) of the RPC call
 	 * and reply headers, and convert both values
@@ -2096,17 +2135,14 @@
 	case -EPROTONOSUPPORT:
 		trace_rpcb_bind_version_err(task);
 		goto retry_timeout;
-	case -ENETDOWN:
-	case -ENETUNREACH:
-		if (task->tk_flags & RPC_TASK_NETUNREACH_FATAL)
-			break;
-		fallthrough;
 	case -ECONNREFUSED:		/* connection problems */
 	case -ECONNRESET:
 	case -ECONNABORTED:
 	case -ENOTCONN:
 	case -EHOSTDOWN:
+	case -ENETDOWN:
 	case -EHOSTUNREACH:
+	case -ENETUNREACH:
 	case -EPIPE:
 		trace_rpcb_unreachable_err(task);
 		if (!RPC_IS_SOFTCONN(task)) {
@@ -2127,7 +2163,7 @@
 retry_timeout:
 	task->tk_status = 0;
 	task->tk_action = call_bind;
-	rpc_check_timeout(task);
+	rpc_check_timeout(task, true);
 }
 
 /*
@@ -2188,22 +2224,19 @@
 
 	task->tk_status = 0;
 	switch (status) {
-	case -ENETDOWN:
-	case -ENETUNREACH:
-		if (task->tk_flags & RPC_TASK_NETUNREACH_FATAL)
-			break;
-		fallthrough;
 	case -ECONNREFUSED:
 	case -ECONNRESET:
 		/* A positive refusal suggests a rebind is needed. */
+		if (RPC_IS_SOFTCONN(task))
+			break;
 		if (clnt->cl_autobind) {
 			rpc_force_rebind(clnt);
-			if (RPC_IS_SOFTCONN(task))
-				break;
 			goto out_retry;
 		}
 		fallthrough;
 	case -ECONNABORTED:
+	case -ENETDOWN:
+	case -ENETUNREACH:
 	case -EHOSTUNREACH:
 	case -EPIPE:
 	case -EPROTO:
@@ -2224,7 +2257,9 @@
 			struct rpc_xprt *saved = task->tk_xprt;
 			struct rpc_xprt_switch *xps;
 
-			xps = rpc_clnt_xprt_switch_get(clnt);
+			rcu_read_lock();
+			xps = xprt_switch_get(rcu_dereference(clnt->cl_xpi.xpi_xpswitch));
+			rcu_read_unlock();
 			if (xps->xps_nxprts > 1) {
 				long value;
 
@@ -2239,7 +2274,7 @@
 			}
 			xprt_switch_put(xps);
 			if (!task->tk_xprt)
-				goto out;
+				return;
 		}
 		goto out_retry;
 	case -ENOBUFS:
@@ -2254,8 +2289,7 @@
 out_retry:
 	/* Check for timeouts before looping back to call_bind */
 	task->tk_action = call_bind;
-out:
-	rpc_check_timeout(task);
+	rpc_check_timeout(task, true);
 }
 
 /*
@@ -2347,7 +2381,7 @@
 		task->tk_status = 0;
 		break;
 	}
-	rpc_check_timeout(task);
+	rpc_check_timeout(task, false);
 }
 
 #if defined(CONFIG_SUNRPC_BACKCHANNEL)
@@ -2455,13 +2489,10 @@
 	trace_rpc_call_status(task);
 	task->tk_status = 0;
 	switch(status) {
-	case -ENETDOWN:
-	case -ENETUNREACH:
-		if (task->tk_flags & RPC_TASK_NETUNREACH_FATAL)
-			goto out_exit;
-		fallthrough;
 	case -EHOSTDOWN:
+	case -ENETDOWN:
 	case -EHOSTUNREACH:
+	case -ENETUNREACH:
 	case -EPERM:
 		if (RPC_IS_SOFTCONN(task))
 			goto out_exit;
@@ -2500,7 +2531,7 @@
 		goto out_exit;
 	}
 	task->tk_action = call_encode;
-	rpc_check_timeout(task);
+	rpc_check_timeout(task, true);
 	return;
 out_exit:
 	rpc_call_rpcerror(task, status);
@@ -2516,15 +2547,20 @@
 }
 
 static void
-rpc_check_timeout(struct rpc_task *task)
+rpc_check_timeout(struct rpc_task *task, bool failover)
 {
 	struct rpc_clnt	*clnt = task->tk_client;
 
 	if (RPC_SIGNALLED(task))
 		return;
 
-	if (xprt_adjust_timeout(task->tk_rqstp) == 0)
+	if (xprt_adjust_timeout(task->tk_rqstp) == 0) {
+#if IS_ENABLED(CONFIG_SUNRPC_ENFS)
+		if (failover)
+			rpc_multipath_ops_failover_handle(task);
+#endif
 		return;
+	}
 
 	trace_rpc_timeout_status(task);
 	task->tk_timeouts++;
@@ -2635,11 +2671,11 @@
 			xprt_conditional_disconnect(req->rq_xprt,
 						    req->rq_connect_cookie);
 		task->tk_action = call_encode;
-		rpc_check_timeout(task);
+		rpc_check_timeout(task, true);
 		break;
 	case -EKEYREJECTED:
 		task->tk_action = call_reserve;
-		rpc_check_timeout(task);
+		rpc_check_timeout(task, true);
 		rpcauth_invalcred(task);
 		/* Ensure we obtain a new XID if we retry! */
 		xprt_release(task);
@@ -2661,8 +2697,12 @@
 	*p++ = req->rq_xid;
 	*p++ = rpc_call;
 	*p++ = cpu_to_be32(RPC_VERSION);
+#if IS_ENABLED(CONFIG_SUNRPC_ENFS)
+	RPC_MULTIPAHT_UPDATE_RPC_PROC(task, p, clnt);
+#else
 	*p++ = cpu_to_be32(clnt->cl_prog);
 	*p++ = cpu_to_be32(clnt->cl_vers);
+#endif
 	*p   = cpu_to_be32(task->tk_msg.rpc_proc->p_proc);
 
 	error = rpcauth_marshcred(task, xdr);
@@ -2700,19 +2740,8 @@
 		goto out_msg_denied;
 
 	error = rpcauth_checkverf(task, xdr);
-	if (error) {
-		struct rpc_cred *cred = task->tk_rqstp->rq_cred;
-
-		if (!test_bit(RPCAUTH_CRED_UPTODATE, &cred->cr_flags)) {
-			rpcauth_invalcred(task);
-			if (!task->tk_cred_retry)
-				goto out_err;
-			task->tk_cred_retry--;
-			trace_rpc__stale_creds(task);
-			return -EKEYREJECTED;
-		}
+	if (error)
 		goto out_verifier;
-	}
 
 	p = xdr_inline_decode(xdr, sizeof(*p));
 	if (!p)
@@ -2763,13 +2792,8 @@
 	case -EPROTONOSUPPORT:
 		goto out_err;
 	case -EACCES:
-		/* possible RPCSEC_GSS out-of-sequence event (RFC2203),
-		 * reset recv state and keep waiting, don't retransmit
-		 */
-		task->tk_rqstp->rq_reply_bytes_recvd = 0;
-		task->tk_status = xprt_request_enqueue_receive(task);
-		task->tk_action = call_transmit_status;
-		return -EBADMSG;
+		/* Re-encode with a fresh cred */
+		fallthrough;
 	default:
 		goto out_garbage;
 	}
@@ -2976,7 +3000,7 @@
 		return -EINVAL;
 	}
 
-	data = kmalloc_obj(*data);
+	data = kmalloc(sizeof(*data), GFP_KERNEL);
 	if (!data)
 		return -ENOMEM;
 	data->xps = xprt_switch_get(xps);
@@ -2998,6 +3022,29 @@
 }
 EXPORT_SYMBOL_GPL(rpc_clnt_test_and_add_xprt);
 
+/* rpc_clnt_test_xprt - Test and add a new transport to a rpc_clnt
+ * @clnt: pointer to struct rpc_clnt
+ * @xprt: pointer struct rpc_xprt
+ * @ops: async operation
+ */
+int rpc_clnt_test_xprt(struct rpc_clnt *clnt, struct rpc_xprt *xprt,
+		       const struct rpc_call_ops *ops, void *data, int flags)
+{
+	struct rpc_cred *cred;
+	struct rpc_task *task;
+
+	cred = authnull_ops.lookup_cred(NULL, NULL, 0);
+	task = rpc_call_null_helper(clnt, xprt, cred,
+				    RPC_TASK_SOFT | RPC_TASK_SOFTCONN | flags,
+				    ops, data);
+	put_rpccred(cred);
+	if (IS_ERR(task))
+		return PTR_ERR(task);
+	rpc_put_task(task);
+	return 1;
+}
+EXPORT_SYMBOL_GPL(rpc_clnt_test_xprt);
+
 static int rpc_clnt_add_xprt_helper(struct rpc_clnt *clnt,
 				    struct rpc_xprt *xprt,
 				    struct rpc_add_xprt_test *data)
@@ -3133,7 +3180,12 @@
 				connect_timeout,
 				reconnect_timeout);
 
+#if IS_ENABLED(CONFIG_SUNRPC_ENFS)
+	rpc_multipath_switch_set_roundrobin(clnt, xps);
+#else
 	rpc_xprt_switch_set_roundrobin(xps);
+#endif
+
 	if (setup) {
 		ret = setup(clnt, xps, xprt, data);
 		if (ret != 0)
@@ -3152,6 +3204,7 @@
 				  struct rpc_xprt *xprt,
 				  struct rpc_add_xprt_test *data)
 {
+	struct rpc_xprt_switch *xps;
 	struct rpc_xprt *main_xprt;
 	int status = 0;
 
@@ -3159,6 +3212,7 @@
 
 	rcu_read_lock();
 	main_xprt = xprt_get(rcu_dereference(clnt->cl_xprt));
+	xps = xprt_switch_get(rcu_dereference(clnt->cl_xpi.xpi_xpswitch));
 	status = rpc_cmp_addr_port((struct sockaddr *)&xprt->addr,
 				   (struct sockaddr *)&main_xprt->addr);
 	rcu_read_unlock();
@@ -3169,6 +3223,7 @@
 	status = rpc_clnt_add_xprt_helper(clnt, xprt, data);
 out:
 	xprt_put(xprt);
+	xprt_switch_put(xps);
 	return status;
 }
 
@@ -3283,27 +3338,34 @@
 }
 EXPORT_SYMBOL_GPL(rpc_set_connect_timeout);
 
+void rpc_clnt_xprt_switch_put(struct rpc_clnt *clnt)
+{
+	rcu_read_lock();
+	xprt_switch_put(rcu_dereference(clnt->cl_xpi.xpi_xpswitch));
+	rcu_read_unlock();
+}
+EXPORT_SYMBOL_GPL(rpc_clnt_xprt_switch_put);
+
 void rpc_clnt_xprt_set_online(struct rpc_clnt *clnt, struct rpc_xprt *xprt)
 {
 	struct rpc_xprt_switch *xps;
 
-	xps = rpc_clnt_xprt_switch_get(clnt);
+	rcu_read_lock();
+	xps = rcu_dereference(clnt->cl_xpi.xpi_xpswitch);
+	rcu_read_unlock();
 	xprt_set_online_locked(xprt, xps);
-	xprt_switch_put(xps);
 }
 
 void rpc_clnt_xprt_switch_add_xprt(struct rpc_clnt *clnt, struct rpc_xprt *xprt)
 {
-	struct rpc_xprt_switch *xps;
-
 	if (rpc_clnt_xprt_switch_has_addr(clnt,
 		(const struct sockaddr *)&xprt->addr)) {
 		return rpc_clnt_xprt_set_online(clnt, xprt);
 	}
-
-	xps = rpc_clnt_xprt_switch_get(clnt);
-	rpc_xprt_switch_add_xprt(xps, xprt);
-	xprt_switch_put(xps);
+	rcu_read_lock();
+	rpc_xprt_switch_add_xprt(rcu_dereference(clnt->cl_xpi.xpi_xpswitch),
+				 xprt);
+	rcu_read_unlock();
 }
 EXPORT_SYMBOL_GPL(rpc_clnt_xprt_switch_add_xprt);
 
@@ -3335,11 +3397,8 @@
 EXPORT_SYMBOL_GPL(rpc_clnt_xprt_switch_has_addr);
 
 #if IS_ENABLED(CONFIG_SUNRPC_DEBUG)
-static void rpc_show_header(struct rpc_clnt *clnt)
+static void rpc_show_header(void)
 {
-	printk(KERN_INFO "clnt[%pISpc] RPC tasks[%d]\n",
-	       (struct sockaddr *)&clnt->cl_xprt->addr,
-	       atomic_read(&clnt->cl_task_count));
 	printk(KERN_INFO "-pid- flgs status -client- --rqstp- "
 		"-timeout ---ops--\n");
 }
@@ -3371,7 +3430,7 @@
 		spin_lock(&clnt->cl_lock);
 		list_for_each_entry(task, &clnt->cl_tasks, tk_task) {
 			if (!header) {
-				rpc_show_header(clnt);
+				rpc_show_header();
 				header++;
 			}
 			rpc_show_task(clnt, task);
